--- 
title: 'single-cell sPLS'
output:
  html_document:
    toc: no
author: "Al J Abadi"
date: "`r Sys.Date()`"
---

```{r, echo=FALSE}
## input/output setup
io <- list()
io$chunk$cache <- !TRUE
io$chunk$echo <-FALSE

knitr::opts_chunk$set(cache=io$chunk$cache, echo=io$chunk$echo,  cache.comment=FALSE, fig.align = 'center',
                      message=FALSE, warning=FALSE, results='hide')
```


# Transcriptome

For cell matching scBSseq:

  * [QC all cells](./01_Transcriptome/01_QC_Transcriptome/01_QC_Transcriptome_all_cells.html)
  * [QC matching cells](./01_Transcriptome/01_QC_Transcriptome/01_QC_Transcriptome.html)
  * [Dimension Reduction](./01_Transcriptome/02_DimensionReduction_Transcriptome/02_DimensionReduction_Transcriptome.html)

# Methylation

  * [Preprocessing](./02_Methylation/01_Preprocessing_Methylation/01_Preprocessing_Methylation.html)
  * [QC](./02_Methylation/02_QC_Methylation/02_QC_Methylation.html)
  * [Dimension Reduction](./02_Methylation/03_DimensionReduction_Methylation/03_DimensionReduction_Methylation.html)

# Integartion

  * [Gene-wise correlations]()
  * [sPLS]()

<!-- The cells are from the Kasumi-1 AML cell line. Cell identifiers are given by a letter (indicating plate row), and a number (indicating plate column), e.g. *_A7_*, *_C10_*. -->


<!-- +  **scRNA-seq Data** -->

<!-- Count matrices which are quality controlled. -->

<!-- +  **scBS-seq Data** -->

<!-- Each .cov file in which each line is a cytosine residue, and the columns are as follows: -->
<!-- <chromosome>  <position>  <strand>  <count methylated>  <count unmethylated>  <C-context> <trinucleotide context> -->
<!-- These reports are produced by Bismark. The functional element is ill defined. Sometimes we look at promoters, sometimes enhancers, sometimes unbiased - occasionally overlapping - 3kb windows to ensure minimisation of missed genomic features. -->

<!-- 40 cells amcth those from RNA-seq. -->

<!-- ```{r,eval=params$eval, echo=params$echo} -->
<!-- ## update as needed. -->
<!-- ## installing the required packages for this analysis, if necessary -->
<!-- required.pkgs = c('mixOmics','SingleCellExperiment','scran','data.table', -->
<!--                   'DESeq2', 'edgeR', 'Rtsne', 'ggplot2', 'ggrepel', 'gridExtra', -->
<!--                   'grid', 'reshape2', -->
<!--                   ## for summarising -->
<!--                   'seqinr', 'stringr', 'doParallel', 'argparse' -->
<!--                   ) -->
<!-- ``` -->

