<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>scMTseq of AML cells - pilot</title>
  <meta name="description" content="Joint methylation and transcriptome sequencing of AML cells">
  <meta name="generator" content="bookdown 0.7 and GitBook 2.6.7">

  <meta property="og:title" content="scMTseq of AML cells - pilot" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Joint methylation and transcriptome sequencing of AML cells" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="scMTseq of AML cells - pilot" />
  
  <meta name="twitter:description" content="Joint methylation and transcriptome sequencing of AML cells" />
  

<meta name="author" content="Al J Abadi">


<meta name="date" content="2018-11-29">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  

<link rel="next" href="scrna-seq-data.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Pilot scMTseq of AML cells</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Data</a></li>
<li class="chapter" data-level="1" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html"><i class="fa fa-check"></i><b>1</b> scRNA-seq Data</a><ul>
<li class="chapter" data-level="1.1" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#load-raw-data"><i class="fa fa-check"></i><b>1.1</b> Load Raw Data</a></li>
<li class="chapter" data-level="1.2" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#preprocessing"><i class="fa fa-check"></i><b>1.2</b> Preprocessing</a><ul>
<li class="chapter" data-level="1.2.1" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#qc-metrics"><i class="fa fa-check"></i><b>1.2.1</b> QC metrics</a></li>
<li class="chapter" data-level="1.2.2" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#cell-qc"><i class="fa fa-check"></i><b>1.2.2</b> Cell QC</a></li>
<li class="chapter" data-level="1.2.3" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#detect-outliers"><i class="fa fa-check"></i><b>1.2.3</b> Detect outliers</a></li>
<li class="chapter" data-level="1.2.4" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#gene-filtering"><i class="fa fa-check"></i><b>1.2.4</b> Gene Filtering</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#normalization"><i class="fa fa-check"></i><b>1.3</b> Normalization</a><ul>
<li class="chapter" data-level="1.3.1" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#scran"><i class="fa fa-check"></i><b>1.3.1</b> scran</a></li>
<li class="chapter" data-level="1.3.2" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#tmm-edger"><i class="fa fa-check"></i><b>1.3.2</b> TMM (edgeR)</a></li>
<li class="chapter" data-level="1.3.3" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#deseq2"><i class="fa fa-check"></i><b>1.3.3</b> DESeq2</a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#dimension-reduction"><i class="fa fa-check"></i><b>1.4</b> Dimension reduction</a><ul>
<li class="chapter" data-level="1.4.1" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#pca"><i class="fa fa-check"></i><b>1.4.1</b> PCA</a></li>
<li class="chapter" data-level="1.4.2" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#hierarchical-clustering"><i class="fa fa-check"></i><b>1.4.2</b> Hierarchical Clustering</a></li>
<li class="chapter" data-level="1.4.3" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#tsne"><i class="fa fa-check"></i><b>1.4.3</b> tSNE</a></li>
<li class="chapter" data-level="1.4.4" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#looking-at-umap"><i class="fa fa-check"></i><b>1.4.4</b> Looking at UMAP</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html"><i class="fa fa-check"></i><b>2</b> scBS-seq Data</a><ul>
<li class="chapter" data-level="2.1" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html#methylation-and-rnaseq"><i class="fa fa-check"></i><b>2.1</b> Methylation and RNAseq</a><ul>
<li class="chapter" data-level="2.1.1" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html#filtering-nas"><i class="fa fa-check"></i><b>2.1.1</b> Filtering NAs</a></li>
<li class="chapter" data-level="2.1.2" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html#hvgs-in-methylation-data"><i class="fa fa-check"></i><b>2.1.2</b> HVGs in Methylation Data</a></li>
<li class="chapter" data-level="2.1.3" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html#spls---gene-body"><i class="fa fa-check"></i><b>2.1.3</b> sPLS - Gene Body</a></li>
<li class="chapter" data-level="2.1.4" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html#spls---promoter"><i class="fa fa-check"></i><b>2.1.4</b> sPLS - Promoter</a></li>
<li class="chapter" data-level="2.1.5" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html#block-spls"><i class="fa fa-check"></i><b>2.1.5</b> Block sPLS</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html#kb-windows"><i class="fa fa-check"></i><b>2.2</b> 3kb Windows</a><ul>
<li class="chapter" data-level="2.2.1" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html#dimension-reduction-1"><i class="fa fa-check"></i><b>2.2.1</b> Dimension Reduction</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html"><i class="fa fa-check"></i><b>3</b> scM&amp;T-seq analysis</a></li>
<li class="divider"></li>
<li><a href="https://github.com/AJABADI/scMTseq-AML/" target="blank">See on GitHub</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">scMTseq of AML cells - pilot</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="header">
<h1 class="title">scMTseq of AML cells - pilot</h1>
<h4 class="author"><em>Al J Abadi</em></h4>
<h4 class="date"><em>2018-11-29</em></h4>
</div>
<div id="data" class="section level1 unnumbered">
<h1>Data</h1>
<p>Data from Dr Heather Lee lab from The University of Newcastle.</p>
<p>The cells are from the Kasumi-1 AML cell line. Cell identifiers are given by a letter (indicating plate row), and a number (indicating plate column), e.g. <em><em>A7</em></em>, <em><em>C10</em></em>.</p>
<ul>
<li><strong>scRNA-seq Data</strong></li>
</ul>
<p>Count matrices which are quality controlled.</p>
<ul>
<li><strong>scBS-seq Data</strong></li>
</ul>
<p>Each .cov file in which each line is a cytosine residue, and the columns are as follows:
<chromosome> <position> <strand> <count methylated> <count unmethylated> <C-context> <trinucleotide context>
These reports are produced by Bismark. The functional element is ill defined. Sometimes we look at promoters, sometimes enhancers, sometimes unbiased - occasionally overlapping - 3kb windows to ensure minimisation of missed genomic features.</p>
<p>40 cells amcth those from RNA-seq.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="co">## update as needed.</span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="co">## installing the required packages for this analysis, if necessary</span></a>
<a class="sourceLine" id="cb1-3" data-line-number="3">required.pkgs =<span class="st"> </span><span class="kw">c</span>(<span class="st">&#39;mixOmics&#39;</span>,<span class="st">&#39;SingleCellExperiment&#39;</span>,<span class="st">&#39;scran&#39;</span>,<span class="st">&#39;data.table&#39;</span>,</a>
<a class="sourceLine" id="cb1-4" data-line-number="4">                  <span class="st">&#39;DESeq2&#39;</span>, <span class="st">&#39;edgeR&#39;</span>, <span class="st">&#39;Rtsne&#39;</span>, <span class="st">&#39;ggplot2&#39;</span>, <span class="st">&#39;ggrepel&#39;</span>, <span class="st">&#39;gridExtra&#39;</span>,</a>
<a class="sourceLine" id="cb1-5" data-line-number="5">                  <span class="st">&#39;grid&#39;</span>, <span class="st">&#39;reshape2&#39;</span>,</a>
<a class="sourceLine" id="cb1-6" data-line-number="6">                  <span class="co">## for summarising</span></a>
<a class="sourceLine" id="cb1-7" data-line-number="7">                  <span class="st">&#39;seqinr&#39;</span>, <span class="st">&#39;stringr&#39;</span>, <span class="st">&#39;doParallel&#39;</span>, <span class="st">&#39;argparse&#39;</span></a>
<a class="sourceLine" id="cb1-8" data-line-number="8">                  )</a></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="co">## make sure Biocmanager is installed</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&#39;BiocManager&#39;</span>, <span class="dt">quietly =</span> T)){</a>
<a class="sourceLine" id="cb2-3" data-line-number="3">  <span class="kw">paste</span>(<span class="st">&#39;Trying to install BiocManager&#39;</span>)</a>
<a class="sourceLine" id="cb2-4" data-line-number="4">  <span class="kw">install.packages</span>(<span class="st">&#39;BiocManager&#39;</span>)</a>
<a class="sourceLine" id="cb2-5" data-line-number="5">}</a></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="co">## package installer function - for those not already installed</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2">package.installer =<span class="st"> </span><span class="cf">function</span>(<span class="dt">pkgs=</span>required.pkgs){</a>
<a class="sourceLine" id="cb3-3" data-line-number="3">  <span class="cf">for</span> (package <span class="cf">in</span> pkgs){</a>
<a class="sourceLine" id="cb3-4" data-line-number="4">    <span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(package, <span class="dt">quietly =</span> T)){</a>
<a class="sourceLine" id="cb3-5" data-line-number="5">  <span class="kw">paste0</span>(<span class="st">&#39;Trying to install &#39;</span>, package)</a>
<a class="sourceLine" id="cb3-6" data-line-number="6">  BiocManager<span class="op">::</span><span class="kw">install</span>(package, <span class="dt">update =</span> F)</a>
<a class="sourceLine" id="cb3-7" data-line-number="7">    }</a>
<a class="sourceLine" id="cb3-8" data-line-number="8">    }</a>
<a class="sourceLine" id="cb3-9" data-line-number="9">}</a>
<a class="sourceLine" id="cb3-10" data-line-number="10"><span class="co">## run function</span></a>
<a class="sourceLine" id="cb3-11" data-line-number="11"><span class="kw">package.installer</span>(required.pkgs)</a></code></pre></div>
<!-- Activate when params are finalised and defined in YAML -->
<!-- ```{r} -->
<!-- ## I/O -->
<!-- io=list() -->
<!--   ## raw RNA data -->
<!--   io$rna_file = params$rna_file ## '../data/scRNA-seq/raw/Kasumi1_scRNAseq_RawCounts.txt' -->
<!--   ## raw single bp met calls folder -->
<!--   io$met_1bp_in = params$met_1bp_in ## '../data/scBS-seq/met/1bp/raw' -->
<!--   ## processed single bp met calls file name -->
<!--   io$met_1bp_out = params$met_1bp_out ## '../data/scBS-seq/met/1bp/parsed/met_data.tsv' -->
<!--   ## whether to use parsed data or (re-)parse -->
<!--   io$use.parsed= params$use.parsed## TRUE -->
<!--   ## 3kb windows data folder -->
<!--   io$met_3kb = params$met_3kb ## '../data/scBS-seq/met/3kbp/raw/methylVariance-w3K-s1K5.csv' -->
<!--   ## .bed files for summarising calls on genomic features -->
<!--   io$bedfiles = params$bedfiles ## '../data/scBS-seq/filt' -->
<!--   io$RData = params$RData ## save rdata -->
<!-- ``` -->
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="co">## check for validity of inputs</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2"><span class="kw">stopifnot</span>(<span class="kw">file.exists</span>(io<span class="op">$</span>rna_file), <span class="op">!</span>io<span class="op">$</span>use.parsed<span class="op">||</span><span class="kw">file.exists</span>(io<span class="op">$</span>met_1bp_out), io<span class="op">$</span>use.parsed<span class="op">||</span><span class="kw">file.exists</span>(io<span class="op">$</span>bedfiles))</a>
<a class="sourceLine" id="cb4-3" data-line-number="3"><span class="co">## for reproducibility</span></a>
<a class="sourceLine" id="cb4-4" data-line-number="4"><span class="kw">set.seed</span>(<span class="dv">1234321</span>)</a></code></pre></div>

</div>
            </section>

          </div>
        </div>
      </div>

<a href="scrna-seq-data.html" class="navigation navigation-next navigation-unique" aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
