<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Chapter 2 scBS-seq Data | scMTseq of AML cells - pilot</title>
  <meta name="description" content="Joint methylation and transcriptome sequencing of AML cells">
  <meta name="generator" content="bookdown  and GitBook 2.6.7">

  <meta property="og:title" content="Chapter 2 scBS-seq Data | scMTseq of AML cells - pilot" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Joint methylation and transcriptome sequencing of AML cells" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 2 scBS-seq Data | scMTseq of AML cells - pilot" />
  
  <meta name="twitter:description" content="Joint methylation and transcriptome sequencing of AML cells" />
  

<meta name="author" content="Al J Abadi">


<meta name="date" content="2019-02-07">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="scrna-seq-data.html">
<link rel="next" href="scmt-seq-analysis.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />







<script type="text/javascript">

// toggle visibility of R source blocks in R Markdown output
function toggle_R() {
  var x = document.getElementsByClassName('r');
  if (x.length == 0) return;
  function toggle_vis(o) {
    var d = o.style.display;
    o.style.display = (d == 'block' || d == '') ? 'none':'block';
  }

  for (i = 0; i < x.length; i++) {
    var y = x[i];
    if (y.tagName.toLowerCase() === 'pre') toggle_vis(y);
  }

    var elem = document.getElementById("myButton1");
    if (elem.value === "Hide Global") elem.value = "Show Global";
    else elem.value = "Hide Global";
}

document.write('<input onclick="toggle_R();" type="button" value="Hide Global" id="myButton1" style="position: absolute; top: 10%; right: 2%; z-index: 200"></input>')

</script>


<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Pilot scMTseq of AML cells</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Data</a></li>
<li class="chapter" data-level="1" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html"><i class="fa fa-check"></i><b>1</b> scRNA-seq Data</a><ul>
<li class="chapter" data-level="1.1" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#load-raw-data"><i class="fa fa-check"></i><b>1.1</b> Load Raw Data</a></li>
<li class="chapter" data-level="1.2" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#preprocessing"><i class="fa fa-check"></i><b>1.2</b> Preprocessing</a><ul>
<li class="chapter" data-level="1.2.1" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#qc-metrics"><i class="fa fa-check"></i><b>1.2.1</b> QC metrics</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#normalisation"><i class="fa fa-check"></i><b>1.3</b> Normalisation</a><ul>
<li class="chapter" data-level="1.3.1" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#scran"><i class="fa fa-check"></i><b>1.3.1</b> scran</a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#dimension-reduction---filtered-genes"><i class="fa fa-check"></i><b>1.4</b> Dimension reduction - filtered genes</a><ul>
<li class="chapter" data-level="1.4.1" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#pca"><i class="fa fa-check"></i><b>1.4.1</b> PCA</a></li>
<li class="chapter" data-level="1.4.2" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#variation"><i class="fa fa-check"></i><b>1.4.2</b> Variation</a></li>
</ul></li>
<li class="chapter" data-level="1.5" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#dimension-reduction---hvgs"><i class="fa fa-check"></i><b>1.5</b> Dimension reduction - HVGs</a><ul>
<li class="chapter" data-level="1.5.1" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#pca-using-hvgs"><i class="fa fa-check"></i><b>1.5.1</b> PCA using HVGS</a></li>
<li class="chapter" data-level="1.5.2" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#tsne-does-not-identify-clear-clusters"><i class="fa fa-check"></i><b>1.5.2</b> tSNE does not identify clear clusters</a></li>
<li class="chapter" data-level="1.5.3" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#umap-does-not-identify-clear-clusters"><i class="fa fa-check"></i><b>1.5.3</b> UMAP does not identify clear clusters</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html"><i class="fa fa-check"></i><b>2</b> scBS-seq Data</a><ul>
<li class="chapter" data-level="2.1" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html#data-1"><i class="fa fa-check"></i><b>2.1</b> Data</a></li>
<li class="chapter" data-level="2.2" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html#create-sce"><i class="fa fa-check"></i><b>2.2</b> Create sce</a></li>
<li class="chapter" data-level="2.3" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html#summary"><i class="fa fa-check"></i><b>2.3</b> Summary</a></li>
<li class="chapter" data-level="2.4" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html#variance"><i class="fa fa-check"></i><b>2.4</b> Variance</a></li>
<li class="chapter" data-level="2.5" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html#dimension-reduction"><i class="fa fa-check"></i><b>2.5</b> Dimension reduction</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html"><i class="fa fa-check"></i><b>3</b> scM&amp;T-seq analysis</a><ul>
<li class="chapter" data-level="3.1" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html#correlation"><i class="fa fa-check"></i><b>3.1</b> Correlation</a><ul>
<li class="chapter" data-level="3.1.1" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html#pearson-correlation-study-of-methylation-rate-and-expression---gene-wise"><i class="fa fa-check"></i><b>3.1.1</b> Pearson correlation study of methylation rate and expression - gene wise</a></li>
<li class="chapter" data-level="3.1.2" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html#does-the-variation-in-methylation-in-genebody-correlates-with-variation-in-expression"><i class="fa fa-check"></i><b>3.1.2</b> Does the variation in methylation in genebody correlates with variation in expression?</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html#canonical-correlation-study"><i class="fa fa-check"></i><b>3.2</b> Canonical correlation study</a><ul>
<li class="chapter" data-level="3.2.1" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html#trascriptome---promoter-methylome"><i class="fa fa-check"></i><b>3.2.1</b> Trascriptome - Promoter Methylome</a></li>
<li class="chapter" data-level="3.2.2" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html#run-using-highly-variable-genes"><i class="fa fa-check"></i><b>3.2.2</b> Run using highly variable genes</a></li>
<li class="chapter" data-level="3.2.3" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html#more-strict-filtering-by-coverage"><i class="fa fa-check"></i><b>3.2.3</b> More strict filtering by coverage</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html#outlier-cells-diagnostics"><i class="fa fa-check"></i><b>3.3</b> Outlier cells diagnostics:</a></li>
<li class="chapter" data-level="3.4" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html#block-spls"><i class="fa fa-check"></i><b>3.4</b> Block sPLS</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/AJABADI/scMTseq-AML/" target="blank">See on GitHub</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">scMTseq of AML cells - pilot</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="scbs-seq-data" class="section level1">
<h1><span class="header-section-number">Chapter 2</span> scBS-seq Data</h1>
<p>40 cells matching RNA-seq data</p>
<div id="data-1" class="section level2">
<h2><span class="header-section-number">2.1</span> Data</h2>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## load the methylation data</span>
met =<span class="st"> </span><span class="kw">fread</span>(io<span class="op">$</span>met_1bp_out)
<span class="kw">unique</span>(met<span class="op">$</span>anno) <span class="co">## all annotations</span>
<span class="co">## look at sample names</span>
<span class="kw">head</span>(<span class="kw">unique</span>(met<span class="op">$</span>sample), <span class="dv">2</span>)
<span class="co">## keep only the well number as the sample name</span>
<span class="co">## the following code is fast but only works if the well names are 2 character long and at the beginning on the string</span>
<span class="co">## if wells such as A12 are present (with 3 or more characters) the function name.cells.by.wells should be called</span>
met<span class="op">$</span>sample =<span class="st"> </span><span class="kw">substr</span>(met<span class="op">$</span>sample,<span class="dv">0</span>,<span class="dv">2</span>)
<span class="kw">length</span>(<span class="kw">unique</span>(met<span class="op">$</span>sample)) <span class="co">## 40</span>
met =met[anno<span class="op">!=</span><span class="st">&quot;intergenic&quot;</span>] <span class="co">## #TODO drop intergenic for now</span></code></pre>
</div>
<div id="create-sce" class="section level2">
<h2><span class="header-section-number">2.2</span> Create sce</h2>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## from each annotation (genebody, promoter etc) in methylation file, create a sce object</span>
<span class="co">## which includes the weighted mean methylation rate (rbar) and weighted methylation variance (vhat)</span>
<span class="co">## plus the coverage of the gene in rowData(sce), filter by coverage and record the changes in dim(sce)</span>
<span class="co">## after filtering in metadata(sce)</span>

<span class="kw">system.time</span>({ <span class="co">## 19 sec</span>
  <span class="cf">for</span> (i <span class="cf">in</span> <span class="kw">unique</span>(met<span class="op">$</span>anno) ) {
    <span class="kw">assign</span>(<span class="kw">paste0</span>(<span class="st">&quot;sce_&quot;</span>,i), <span class="kw">met_sce</span>(<span class="dt">met=</span>met, <span class="dt">annot =</span> i, <span class="dt">cov =</span> params<span class="op">$</span>cov.cutoff))
  }
})</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">load</span>(<span class="st">&#39;RData/met_sces.RData&#39;</span>)</code></pre>
</div>
<div id="summary" class="section level2">
<h2><span class="header-section-number">2.3</span> Summary</h2>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## get all annotations</span>
annots =<span class="st"> </span><span class="kw">grep</span>(<span class="st">&quot;sce_&quot;</span>, <span class="kw">ls</span>(), <span class="dt">value=</span><span class="ot">TRUE</span>)
<span class="co">## tabulate metadata</span>
<span class="kw">lapply</span>(annots, <span class="cf">function</span>(x){<span class="kw">metadata</span>(<span class="kw">get</span>(x))<span class="op">$</span>gene_stats}) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">lapply</span>(., cbind) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as.data.frame</span>() <span class="op">%&gt;%</span><span class="st"> </span>kableExtra<span class="op">::</span><span class="kw">kable</span>(., <span class="dt">caption =</span> <span class="st">&quot;Number of genes in each dataset before and after filtering by coverage&quot;</span>)</code></pre>
</div>
<div id="variance" class="section level2">
<h2><span class="header-section-number">2.4</span> Variance</h2>
<pre class="sourceCode r"><code class="sourceCode r">rv_plot =<span class="st"> </span><span class="cf">function</span>(<span class="dt">sce =</span> sce_genebody, <span class="dt">annot =</span> <span class="st">&quot;Gene Body Methylation&quot;</span>){
  rowdat =<span class="st"> </span>sce <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">rowData</span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as.data.frame</span>()
  <span class="kw">ggplot</span>(rowdat, <span class="kw">aes</span>(<span class="dt">x=</span>rbar, <span class="dt">y=</span>vhat, <span class="dt">col =</span> cov)) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span><span class="kw">guides</span>(<span class="dt">col=</span><span class="kw">guide_legend</span>(<span class="st">&quot;Coverage&quot;</span>)) <span class="op">+</span>
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">x=</span><span class="kw">TeX</span>(<span class="st">&quot;$</span><span class="ch">\\</span><span class="st">bar{r}$&quot;</span>), <span class="dt">y=</span><span class="kw">TeX</span>(<span class="st">&quot;$</span><span class="ch">\\</span><span class="st">hat{</span><span class="ch">\\</span><span class="st">nu}$&quot;</span>),<span class="dt">title =</span> annot)
}</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## plot all vhat ~ rbar</span>
<span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="kw">ceiling</span>(<span class="kw">length</span>(annots)<span class="op">/</span><span class="dv">2</span>),<span class="dv">2</span>))
<span class="kw">lapply</span>(annots, <span class="cf">function</span>(x){<span class="kw">rv_plot</span>(<span class="kw">get</span>(x), x)})</code></pre>
<p><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-40-1.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-40-2.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-40-3.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-40-4.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="dimension-reduction" class="section level2">
<h2><span class="header-section-number">2.5</span> Dimension reduction</h2>
<p>Because of large number of NA’s, dimension reduction cannot be trusted for now.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## pca function</span>
<span class="co">## sce: sce with &#39;rates&#39; assay</span>
<span class="co">## ncomp: integer, number of comps to get</span>
<span class="co">## hvgs: NULL or number of hvgs to keep based on rowData(sce)$vhat</span>
pca_met =<span class="st"> </span><span class="cf">function</span>(<span class="dt">sce =</span> <span class="st">&quot;sce_genebody&quot;</span>, <span class="dt">ncomp =</span> <span class="dv">2</span>, <span class="dt">n_hvgs =</span> <span class="ot">NULL</span>){
  rates =<span class="st"> </span><span class="kw">assay</span>(<span class="kw">get</span>(sce), <span class="st">&quot;rates&quot;</span>)
  <span class="cf">if</span> (<span class="op">!</span><span class="kw">is.null</span>(n_hvgs)){
    <span class="cf">if</span>(n_hvgs<span class="op">&gt;=</span><span class="kw">dim</span>(<span class="kw">get</span>(sce))[<span class="dv">1</span>]){
      <span class="kw">message</span>(<span class="st">&quot;number of genes &lt;= hvgs; all genes used&quot;</span>)
    } <span class="cf">else</span>{
      rates =<span class="st"> </span>rates[<span class="dv">1</span><span class="op">:</span>n_hvgs,]
    }
  }
  out =<span class="st"> </span><span class="kw">pca</span>(<span class="kw">t</span>(rates), <span class="dt">ncomp =</span> ncomp)
  out<span class="op">$</span>anno =<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;sce_&quot;</span>,<span class="st">&quot;&quot;</span>, sce)
  <span class="kw">return</span>(out)
}</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">n_hvgs =<span class="st"> </span>params<span class="op">$</span>hvgs_pca
<span class="co">## maximum number of components reaches for some comps</span>
<span class="kw">system.time</span>({ <span class="co">## 562 sec for ncomp=3, nhvgs =2000</span>
  pca_out =<span class="st"> </span><span class="kw">lapply</span>(annots, <span class="cf">function</span>(x){<span class="kw">pca_met</span>(x,<span class="dt">ncomp =</span> <span class="dv">3</span>, <span class="dt">n_hvgs =</span>n_hvgs)})
  <span class="kw">save</span>(pca_out, <span class="dt">file=</span><span class="st">&quot;RData/pca_out.RData&quot;</span>)
})</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">load</span>(<span class="st">&quot;RData/pca_out.RData&quot;</span>)</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">lapply</span>(pca_out, <span class="cf">function</span>(x){<span class="kw">pca.grid</span>(x, <span class="dt">top =</span> x<span class="op">$</span>anno, <span class="dt">point.label =</span> <span class="ot">TRUE</span>)})</code></pre>
<p><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-44-1.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-44-2.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-44-3.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-44-4.png" width="672" style="display: block; margin: auto;" />
## Weighted PCA?</p>
<!--
impute using DeepCpG and plot the heatmaps:


```r
## hclust and heatmap for genebody methylation
plotHeatmap(met.gene.sce.filt_NAs.hvg, features = rownames(met.gene.sce.filt_NAs.hvg), exprs_values = "counts", zlim = c(0,100),
            clutering_distance_rows = 'correlation', clutering_distance_cols = 'correlation')
```
-->

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="scrna-seq-data.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="scmt-seq-analysis.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
