<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Chapter 3 scM&amp;T-seq analysis | scMTseq of AML cells - pilot</title>
  <meta name="description" content="Joint methylation and transcriptome sequencing of AML cells">
  <meta name="generator" content="bookdown  and GitBook 2.6.7">

  <meta property="og:title" content="Chapter 3 scM&amp;T-seq analysis | scMTseq of AML cells - pilot" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Joint methylation and transcriptome sequencing of AML cells" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 3 scM&amp;T-seq analysis | scMTseq of AML cells - pilot" />
  
  <meta name="twitter:description" content="Joint methylation and transcriptome sequencing of AML cells" />
  

<meta name="author" content="Al J Abadi">


<meta name="date" content="2019-02-11">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="scbs-seq-data.html">

<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />







<script type="text/javascript">

// toggle visibility of R source blocks in R Markdown output
function toggle_R() {
  var x = document.getElementsByClassName('r');
  if (x.length == 0) return;
  function toggle_vis(o) {
    var d = o.style.display;
    o.style.display = (d == 'block' || d == '') ? 'none':'block';
  }

  for (i = 0; i < x.length; i++) {
    var y = x[i];
    if (y.tagName.toLowerCase() === 'pre') toggle_vis(y);
  }

    var elem = document.getElementById("myButton1");
    if (elem.value === "Hide Global") elem.value = "Show Global";
    else elem.value = "Hide Global";
}

document.write('<input onclick="toggle_R();" type="button" value="Hide Global" id="myButton1" style="position: absolute; top: 10%; right: 2%; z-index: 200"></input>')

</script>


<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Pilot scMTseq of AML cells</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Data</a></li>
<li class="chapter" data-level="1" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html"><i class="fa fa-check"></i><b>1</b> scRNA-seq Data</a><ul>
<li class="chapter" data-level="1.1" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#load-raw-data"><i class="fa fa-check"></i><b>1.1</b> Load Raw Data</a></li>
<li class="chapter" data-level="1.2" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#preprocessing"><i class="fa fa-check"></i><b>1.2</b> Preprocessing</a><ul>
<li class="chapter" data-level="1.2.1" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#qc-metrics"><i class="fa fa-check"></i><b>1.2.1</b> QC metrics</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#normalisation"><i class="fa fa-check"></i><b>1.3</b> Normalisation</a><ul>
<li class="chapter" data-level="1.3.1" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#scran"><i class="fa fa-check"></i><b>1.3.1</b> scran</a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#dimension-reduction---filtered-genes"><i class="fa fa-check"></i><b>1.4</b> Dimension reduction - filtered genes</a><ul>
<li class="chapter" data-level="1.4.1" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#pca"><i class="fa fa-check"></i><b>1.4.1</b> PCA</a></li>
<li class="chapter" data-level="1.4.2" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#variation"><i class="fa fa-check"></i><b>1.4.2</b> Variation</a></li>
</ul></li>
<li class="chapter" data-level="1.5" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#dimension-reduction---hvgs"><i class="fa fa-check"></i><b>1.5</b> Dimension reduction - HVGs</a><ul>
<li class="chapter" data-level="1.5.1" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#pca-using-hvgs"><i class="fa fa-check"></i><b>1.5.1</b> PCA using HVGS</a></li>
<li class="chapter" data-level="1.5.2" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#tsne-does-not-identify-clear-clusters"><i class="fa fa-check"></i><b>1.5.2</b> tSNE does not identify clear clusters</a></li>
<li class="chapter" data-level="1.5.3" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#umap-does-not-identify-clear-clusters"><i class="fa fa-check"></i><b>1.5.3</b> UMAP does not identify clear clusters</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html"><i class="fa fa-check"></i><b>2</b> scBS-seq Data</a><ul>
<li class="chapter" data-level="2.1" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html#data-1"><i class="fa fa-check"></i><b>2.1</b> Data</a></li>
<li class="chapter" data-level="2.2" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html#create-sce"><i class="fa fa-check"></i><b>2.2</b> Create sce</a></li>
<li class="chapter" data-level="2.3" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html#summary"><i class="fa fa-check"></i><b>2.3</b> Summary</a></li>
<li class="chapter" data-level="2.4" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html#variance"><i class="fa fa-check"></i><b>2.4</b> Variance</a></li>
<li class="chapter" data-level="2.5" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html#dimension-reduction"><i class="fa fa-check"></i><b>2.5</b> Dimension reduction</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html"><i class="fa fa-check"></i><b>3</b> scM&amp;T-seq analysis</a><ul>
<li class="chapter" data-level="3.1" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html#correlation"><i class="fa fa-check"></i><b>3.1</b> Correlation</a><ul>
<li class="chapter" data-level="3.1.1" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html#pearson-correlation-study-of-methylation-rate-and-expression---gene-wise"><i class="fa fa-check"></i><b>3.1.1</b> Pearson correlation study of methylation rate and expression - gene wise</a></li>
<li class="chapter" data-level="3.1.2" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html#does-the-variation-in-methylation-in-genebody-correlates-with-variation-in-expression"><i class="fa fa-check"></i><b>3.1.2</b> Does the variation in methylation in genebody correlates with variation in expression?</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html#canonical-correlation-study"><i class="fa fa-check"></i><b>3.2</b> Canonical correlation study</a><ul>
<li class="chapter" data-level="3.2.1" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html#trascriptome---promoter-methylome"><i class="fa fa-check"></i><b>3.2.1</b> Trascriptome - Promoter Methylome</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/AJABADI/scMTseq-AML/" target="blank">See on GitHub</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">scMTseq of AML cells - pilot</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="scmt-seq-analysis" class="section level1">
<h1><span class="header-section-number">Chapter 3</span> scM&amp;T-seq analysis</h1>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">load</span>(<span class="kw">file.path</span>(params<span class="op">$</span>RData,<span class="st">&#39;rna_qc.RData&#39;</span>))
<span class="kw">load</span>(<span class="kw">file.path</span>(params<span class="op">$</span>RData,<span class="st">&#39;met_sces.RData&#39;</span>))</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## find the common cells in both studies</span>
cell.intersect =<span class="st"> </span><span class="kw">Reduce</span>(intersect, <span class="kw">list</span>(<span class="kw">colnames</span>(sce_genebody), <span class="kw">colnames</span>(rna_qc)))
<span class="kw">length</span>(cell.intersect)

<span class="co">## keep only matching cells</span>
rna_match =<span class="st"> </span>rna_qc[,<span class="kw">colnames</span>(rna_qc) <span class="op">%in%</span><span class="st"> </span>cell.intersect]</code></pre>
<div id="correlation" class="section level2">
<h2><span class="header-section-number">3.1</span> Correlation</h2>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## gene-wise pearson correlation of methylation and expression for top `top_prop` proportion of</span>
<span class="co">## common genes b/w the two datasets based on their weighted variance in methylation datasets</span>
<span class="co">## stored in rowData(met_sces)</span>
<span class="co">## plus</span>
<span class="co">## the global correlation of biological and total variance b/w meth. and expression for all genes in metadata(met_sce)</span>

pearson_runtime =<span class="st"> </span><span class="kw">system.time</span>({ <span class="co">## 90 sec</span>
  met_sces =<span class="st"> </span><span class="kw">grep</span>(<span class="st">&#39;sce_&#39;</span>, <span class="kw">ls</span>(), <span class="dt">value=</span><span class="ot">TRUE</span>)
  <span class="cf">for</span> (i <span class="cf">in</span> met_sces){
    <span class="kw">assign</span>(i, <span class="kw">corr_mt</span>(rna_match, <span class="kw">get</span>(i), <span class="dt">top_prop =</span> <span class="fl">0.3</span>) )
  }
})
pearson_runtime <span class="op">%&lt;&gt;%</span><span class="st"> </span>.[<span class="dv">1</span>] <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">unname</span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">round</span>()
<span class="kw">save</span>(<span class="dt">list =</span> <span class="kw">c</span>(met_sces, <span class="st">&#39;pearson_runtime&#39;</span>), <span class="dt">file =</span> <span class="kw">file.path</span>(params<span class="op">$</span>RData,<span class="st">&#39;met_sces.RData&#39;</span>))</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">met_sces =<span class="st"> </span><span class="kw">grep</span>(<span class="st">&#39;sce_&#39;</span>, <span class="kw">ls</span>(), <span class="dt">value=</span><span class="ot">TRUE</span>)
<span class="kw">load</span>(<span class="kw">file.path</span>(params<span class="op">$</span>RData,<span class="st">&#39;met_sces.RData&#39;</span>))</code></pre>
<div id="pearson-correlation-study-of-methylation-rate-and-expression---gene-wise" class="section level3">
<h3><span class="header-section-number">3.1.1</span> Pearson correlation study of methylation rate and expression - gene wise</h3>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## function to create boxplot of pearson correlations of M&amp;T</span>
<span class="co">## for all sces</span>
boxPlot =<span class="st"> </span><span class="cf">function</span>(<span class="dt">sces =</span> met_sces, <span class="dt">n_top =</span> <span class="ot">NULL</span>, <span class="dt">subTitle =</span> <span class="st">&#39;top variable genes based on $</span><span class="ch">\\</span><span class="st">hat{</span><span class="ch">\\</span><span class="st">nu}$&#39;</span> ){
  factor_sces =<span class="st"> </span><span class="ot">NULL</span>
  corr =<span class="st"> </span><span class="ot">NULL</span>
  <span class="cf">for</span> (i <span class="cf">in</span> sces){
    
    n =<span class="st"> </span><span class="kw">ifelse</span>(<span class="kw">is.null</span>(n_top),
               <span class="kw">dim</span>(<span class="kw">get</span>(i))[<span class="dv">1</span>], 
               <span class="kw">min</span>(n_top, <span class="kw">dim</span>(<span class="kw">get</span>(i))[<span class="dv">1</span>]))
    
    factor_sces =<span class="st"> </span><span class="kw">c</span>(factor_sces, <span class="kw">rep</span>(i, n))
    corr =<span class="st"> </span><span class="kw">c</span>(corr,<span class="kw">rowData</span>(<span class="kw">get</span>(i)[<span class="dv">1</span><span class="op">:</span>n,])<span class="op">$</span>corr_rna)
  }
  
  p =<span class="st"> </span><span class="kw">ggplot</span>(<span class="dt">data =</span> <span class="ot">NULL</span>, <span class="kw">aes</span>(<span class="dt">x =</span> factor_sces , <span class="dt">y=</span>corr, <span class="dt">fill =</span> <span class="kw">factor</span>(factor_sces))) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">geom_boxplot</span>(<span class="dt">na.rm =</span> <span class="ot">TRUE</span>) <span class="op">+</span><span class="st"> </span><span class="kw">coord_flip</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&#39;&#39;</span>, <span class="dt">y =</span> <span class="st">&#39;Pearson correlation&#39;</span>, 
         <span class="dt">title =</span> <span class="st">&#39;Gene-wise correlation of methylation and transcription&#39;</span>, <span class="dt">subtitle =</span> <span class="kw">TeX</span>(subTitle)) <span class="op">+</span>
<span class="st">    </span><span class="kw">guides</span>(<span class="dt">fill=</span><span class="ot">FALSE</span>)
  <span class="kw">return</span>(p)
  }</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## Pearson correlation boxplots</span>
<span class="kw">boxPlot</span>(<span class="dt">sces =</span> met_sces, <span class="dt">n_top =</span> <span class="ot">NULL</span>, <span class="dt">subTitle =</span> <span class="st">&#39;top 30% variable genes based on $</span><span class="ch">\\</span><span class="st">hat{</span><span class="ch">\\</span><span class="st">nu}$&#39;</span> )</code></pre>
<p><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-52-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="does-the-variation-in-methylation-in-genebody-correlates-with-variation-in-expression" class="section level3">
<h3><span class="header-section-number">3.1.2</span> Does the variation in methylation in genebody correlates with variation in expression?</h3>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## initialise a data.frame to store the pearson correlation of total and biological variance...</span>
<span class="co">## b/w expression and each methylation dataset</span>
var_corr =<span class="st"> </span><span class="kw">data.frame</span>(<span class="kw">matrix</span>(<span class="dt">nrow =</span> <span class="kw">length</span>(met_sces),<span class="dt">ncol =</span> <span class="dv">2</span>),
                      <span class="dt">row.names =</span> met_sces )
<span class="kw">colnames</span>(var_corr) =<span class="st"> </span><span class="kw">c</span>(<span class="st">&#39;total&#39;</span>, <span class="st">&#39;biological&#39;</span>)

<span class="co">## calculate them</span>
<span class="cf">for</span> (i <span class="cf">in</span> met_sces){
  var_corr[i,] =<span class="st"> </span><span class="kw">c</span>(<span class="kw">metadata</span>(<span class="kw">get</span>(i))<span class="op">$</span>corr_var_tot_wtd_by_cov[<span class="dv">1</span>],
                   <span class="kw">metadata</span>(<span class="kw">get</span>(i))<span class="op">$</span>corr_var_bio_wtd_by_cov[<span class="dv">1</span>])
}
<span class="co">## plot</span>
<span class="kw">ggplot</span>() <span class="op">+</span><span class="st"> </span><span class="kw">geom_col</span>(<span class="kw">aes</span>(<span class="kw">factor</span>(<span class="kw">rownames</span>(var_corr)),var_corr<span class="op">$</span>total, <span class="dt">fill =</span> <span class="kw">factor</span>(<span class="kw">rownames</span>(var_corr))), <span class="dt">shape=</span><span class="dv">23</span>, <span class="dt">size=</span><span class="dv">4</span>) <span class="op">+</span><span class="st"> </span><span class="kw">guides</span>(<span class="dt">fill=</span><span class="ot">FALSE</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&#39;Genomic Context&#39;</span>, <span class="dt">y=</span> <span class="st">&#39;Pearson correlation&#39;</span>, <span class="dt">title =</span> <span class="st">&#39;Correlation of total transcrption variance </span><span class="ch">\n</span><span class="st"> and weighted methylation variance&#39;</span>) <span class="op">+</span><span class="st"> </span><span class="kw">coord_cartesian</span>(<span class="dt">ylim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="fl">0.3</span>, <span class="fl">0.3</span>))</code></pre>
<p><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-53-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>() <span class="op">+</span><span class="st"> </span><span class="kw">geom_col</span>(<span class="kw">aes</span>(<span class="kw">factor</span>(<span class="kw">rownames</span>(var_corr)),var_corr<span class="op">$</span>bio, <span class="dt">fill =</span> <span class="kw">factor</span>(<span class="kw">rownames</span>(var_corr))), <span class="dt">shape=</span><span class="dv">23</span>, <span class="dt">size=</span><span class="dv">4</span>) <span class="op">+</span><span class="st"> </span><span class="kw">guides</span>(<span class="dt">fill=</span><span class="ot">FALSE</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&#39;Genomic Context&#39;</span>, <span class="dt">y=</span> <span class="st">&#39;Pearson correlation&#39;</span>, <span class="dt">title =</span> <span class="st">&#39;Correlation of decomposed biological transcrption variance </span><span class="ch">\n</span><span class="st"> and weighted methylation variance&#39;</span>)<span class="op">+</span><span class="st"> </span><span class="kw">coord_cartesian</span>(<span class="dt">ylim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="fl">0.3</span>, <span class="fl">0.3</span>))</code></pre>
<p><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-53-2.png" width="672" style="display: block; margin: auto;" />
Non-CGI promoters have the highest correlation magnititude and show positive correlation when total variance is used, and negative when biological is considered.</p>
</div>
</div>
<div id="canonical-correlation-study" class="section level2">
<h2><span class="header-section-number">3.2</span> Canonical correlation study</h2>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## dataset parameters for plot</span>
<span class="co">## first one will be taken as Y, the rest X&#39;s</span>
<span class="co">## all that are in the sPLS list + RNA</span>
par_plot =<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">row.names =</span> <span class="kw">c</span>(<span class="st">&#39;rna_match&#39;</span>, <span class="st">&#39;sce_genebody&#39;</span>, <span class="st">&#39;sce_prom&#39;</span>, <span class="st">&#39;sce_prom_cgi&#39;</span>, <span class="st">&#39;sce_prom_noncgi&#39;</span> ),
                      <span class="dt">label =</span> <span class="kw">c</span>(<span class="st">&#39;Transcriptome&#39;</span>,<span class="st">&#39;Genebody&#39;</span>, <span class="st">&#39;Promoter&#39;</span>, <span class="st">&#39;Promoter (CGI)&#39;</span>, <span class="st">&#39;Promoter (non-CGI)&#39;</span>),
                      <span class="dt">shape =</span><span class="kw">c</span>(<span class="dv">17</span>, <span class="dv">19</span>, <span class="dv">15</span>, <span class="dv">18</span>, <span class="dv">8</span>), <span class="co">## shapes for corr circ plots</span>
                      <span class="dt">col =</span> <span class="kw">color.mixo</span>(<span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">4</span>, <span class="dv">5</span>, <span class="dv">8</span>))) <span class="co">## colors for corr circ plots</span></code></pre>
<div id="trascriptome---promoter-methylome" class="section level3">
<h3><span class="header-section-number">3.2.1</span> Trascriptome - Promoter Methylome</h3>
<!--PICKUP-->
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## spls for all methylation data against transcriptome</span>
<span class="co">## function for spls using top x% variable genes from each dataset</span>
top_pct_spls =<span class="st"> </span><span class="cf">function</span>(<span class="dt">Top_pct =</span> top_pct, <span class="co">## portion of top genes</span>
                        <span class="dt">Spls_name =</span> <span class="ot">NULL</span>,<span class="co">## name of the spls list</span>
                        <span class="dt">NZV =</span> nzv ){ <span class="co">## near-zero-variance in data?</span>
  <span class="cf">if</span>(<span class="kw">is.null</span>(Spls_name)) {Spls_name =<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&#39;spls_all_top_&#39;</span>, Top_pct<span class="op">*</span><span class="dv">100</span>,<span class="st">&quot;pct&quot;</span>)}
  
  <span class="kw">assign</span>(Spls_name,
       <span class="kw">spls_wrapper</span>(<span class="dt">x=</span><span class="dv">2</span><span class="op">:</span><span class="dv">5</span>, <span class="dt">block =</span> <span class="ot">FALSE</span>, <span class="dt">top_x =</span> Top_pct, <span class="dt">top_y =</span> Top_pct, <span class="dt">near_zer_var =</span> NZV))
  <span class="kw">save</span>(<span class="dt">list=</span><span class="kw">c</span>(Spls_name), <span class="dt">file=</span>stringr<span class="op">::</span><span class="kw">str_c</span>(params<span class="op">$</span>RData,<span class="st">&#39;/&#39;</span>,Spls_name,<span class="st">&#39;.RData&#39;</span> ))
  <span class="kw">return</span>(Spls_name)
}

<span class="co">## function to create spls object name</span>
splsName =<span class="st"> </span><span class="cf">function</span>(<span class="dt">Top_pct =</span> top_pct) {spls_name =<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&#39;spls_all_top_&#39;</span>, Top_pct<span class="op">*</span><span class="dv">100</span>,<span class="st">&quot;pct&quot;</span>)}</code></pre>
<div id="all-genes" class="section level4">
<h4><span class="header-section-number">3.2.1.1</span> All genes</h4>
<pre class="sourceCode r"><code class="sourceCode r">top_pct =<span class="st"> </span><span class="dv">1</span> <span class="co">## all genes</span>
nzv =<span class="st"> </span><span class="ot">TRUE</span>

spls_name =<span class="st"> </span><span class="kw">splsName</span>(top_pct)
<span class="kw">top_pct_spls</span>(top_pct,<span class="dt">Spls_name =</span> spls_name, <span class="dt">NZV =</span> nzv)</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">source</span>(<span class="st">&#39;utils/spls_plot.R&#39;</span>)
<span class="kw">spls_plot</span>(<span class="kw">get</span>(spls_name), <span class="dt">pars =</span> par_plot)</code></pre>
<p><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-58-1.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-58-2.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-58-3.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-58-4.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-58-5.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-58-6.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-58-7.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-58-8.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-58-9.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-58-10.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-58-11.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-58-12.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-58-13.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-58-14.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-58-15.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-58-16.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">source</span>(<span class="st">&#39;utils/spls_plot.R&#39;</span>)
<span class="kw">png</span>(<span class="dt">filename =</span> <span class="st">&#39;../output/png/plotVar_gbody_transcriptome_all_genes.png&#39;</span>,<span class="dt">height =</span> <span class="dv">700</span>, <span class="dt">width =</span> <span class="dv">700</span><span class="op">/</span><span class="fl">0.8</span> )
<span class="kw">spls_plot</span>(spls_all, <span class="dt">pars =</span> par_plot[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>,],<span class="dt">plots =</span> <span class="st">&#39;plotVar&#39;</span>)
<span class="kw">dev.off</span>()</code></pre>
</div>
<div id="using-top-50-genes-from-each-dataset" class="section level4">
<h4><span class="header-section-number">3.2.1.2</span> using top 50% genes from each dataset:</h4>
<pre class="sourceCode r"><code class="sourceCode r">top_pct =<span class="st"> </span><span class="fl">0.5</span> <span class="co">## all genes</span>
nzv =<span class="st"> </span><span class="ot">TRUE</span>

spls_name =<span class="st"> </span><span class="kw">splsName</span>(top_pct)
<span class="kw">top_pct_spls</span>(top_pct,<span class="dt">Spls_name =</span> spls_name, <span class="dt">NZV =</span> nzv)</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">spls_plot</span>(<span class="kw">get</span>(spls_name), <span class="dt">pars =</span> par_plot)</code></pre>
<p><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-62-1.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-62-2.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-62-3.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-62-4.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-62-5.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-62-6.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-62-7.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-62-8.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-62-9.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-62-10.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-62-11.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-62-12.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-62-13.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-62-14.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-62-15.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-62-16.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">source</span>(<span class="st">&#39;utils/spls_plot.R&#39;</span>)
png_name =<span class="st"> </span><span class="kw">str_c</span>(<span class="st">&#39;../output/png/plotVar_gbody_transcriptome_top_&#39;</span>, top_pct<span class="op">*</span><span class="dv">100</span>, <span class="st">&#39;_pct_genes_from_each.png&#39;</span> )
<span class="kw">png</span>(<span class="dt">filename =</span> png_name, <span class="dt">height =</span> <span class="dv">700</span>, <span class="dt">width =</span> <span class="dv">700</span><span class="op">/</span><span class="fl">0.8</span>)
<span class="kw">spls_plot</span>(<span class="kw">get</span>(spls_name), <span class="dt">pars =</span> par_plot[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>,],<span class="dt">plots =</span> <span class="st">&#39;plotVar&#39;</span>)
<span class="kw">dev.off</span>()</code></pre>
</div>
<div id="using-top-30-genes-from-each-dataset" class="section level4">
<h4><span class="header-section-number">3.2.1.3</span> using top 30% genes from each dataset:</h4>
<pre class="sourceCode r"><code class="sourceCode r">top_pct =<span class="st"> </span><span class="fl">0.3</span> <span class="co">## all genes</span>
nzv =<span class="st"> </span><span class="ot">TRUE</span>

spls_name =<span class="st"> </span><span class="kw">splsName</span>(top_pct)
<span class="kw">top_pct_spls</span>(top_pct,<span class="dt">Spls_name =</span> spls_name, <span class="dt">NZV =</span> nzv)</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">spls_plot</span>(<span class="kw">get</span>(spls_name), <span class="dt">pars =</span> par_plot)</code></pre>
<p><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-66-1.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-66-2.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-66-3.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-66-4.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-66-5.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-66-6.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-66-7.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-66-8.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-66-9.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-66-10.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-66-11.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-66-12.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-66-13.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-66-14.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-66-15.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-66-16.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">source</span>(<span class="st">&#39;utils/spls_plot.R&#39;</span>)
png_name =<span class="st"> </span><span class="kw">str_c</span>(<span class="st">&#39;../output/png/plotVar_gbody_transcriptome_top_&#39;</span>, top_pct<span class="op">*</span><span class="dv">100</span>, <span class="st">&#39;_pct_genes_from_each.png&#39;</span> )
<span class="kw">png</span>(<span class="dt">filename =</span> png_name, <span class="dt">height =</span> <span class="dv">700</span>, <span class="dt">width =</span> <span class="dv">700</span><span class="op">/</span><span class="fl">0.8</span>)
<span class="kw">spls_plot</span>(<span class="kw">get</span>(spls_name), <span class="dt">pars =</span> par_plot[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>,],<span class="dt">plots =</span> <span class="st">&#39;plotVar&#39;</span>)
<span class="kw">dev.off</span>()</code></pre>

<div id="refs" class="references">
<div>
<p>Lun, Aaron T.L., Davis J. McCarthy, and John C. Marioni. 2018. “A Step-by-Step Workflow for Low-Level Analysis of Single-Cell Rna-Seq Data with Bioconductor.” <em>F1000Research</em> 5 (2122). <a href="https://doi.org/10.12688/f1000research.9501.2">https://doi.org/10.12688/f1000research.9501.2</a>.</p>
</div>
</div>
</div>
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="scbs-seq-data.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
