<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Chapter 3 scM&amp;T-seq analysis | scMTseq of AML cells - pilot</title>
  <meta name="description" content="Joint methylation and transcriptome sequencing of AML cells">
  <meta name="generator" content="bookdown  and GitBook 2.6.7">

  <meta property="og:title" content="Chapter 3 scM&amp;T-seq analysis | scMTseq of AML cells - pilot" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Joint methylation and transcriptome sequencing of AML cells" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 3 scM&amp;T-seq analysis | scMTseq of AML cells - pilot" />
  
  <meta name="twitter:description" content="Joint methylation and transcriptome sequencing of AML cells" />
  

<meta name="author" content="Al J Abadi">


<meta name="date" content="2019-02-07">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="scbs-seq-data.html">

<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />







<script type="text/javascript">

// toggle visibility of R source blocks in R Markdown output
function toggle_R() {
  var x = document.getElementsByClassName('r');
  if (x.length == 0) return;
  function toggle_vis(o) {
    var d = o.style.display;
    o.style.display = (d == 'block' || d == '') ? 'none':'block';
  }

  for (i = 0; i < x.length; i++) {
    var y = x[i];
    if (y.tagName.toLowerCase() === 'pre') toggle_vis(y);
  }

    var elem = document.getElementById("myButton1");
    if (elem.value === "Hide Global") elem.value = "Show Global";
    else elem.value = "Hide Global";
}

document.write('<input onclick="toggle_R();" type="button" value="Hide Global" id="myButton1" style="position: absolute; top: 10%; right: 2%; z-index: 200"></input>')

</script>


<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Pilot scMTseq of AML cells</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Data</a></li>
<li class="chapter" data-level="1" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html"><i class="fa fa-check"></i><b>1</b> scRNA-seq Data</a><ul>
<li class="chapter" data-level="1.1" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#load-raw-data"><i class="fa fa-check"></i><b>1.1</b> Load Raw Data</a></li>
<li class="chapter" data-level="1.2" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#preprocessing"><i class="fa fa-check"></i><b>1.2</b> Preprocessing</a><ul>
<li class="chapter" data-level="1.2.1" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#qc-metrics"><i class="fa fa-check"></i><b>1.2.1</b> QC metrics</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#normalisation"><i class="fa fa-check"></i><b>1.3</b> Normalisation</a><ul>
<li class="chapter" data-level="1.3.1" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#scran"><i class="fa fa-check"></i><b>1.3.1</b> scran</a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#dimension-reduction---filtered-genes"><i class="fa fa-check"></i><b>1.4</b> Dimension reduction - filtered genes</a><ul>
<li class="chapter" data-level="1.4.1" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#pca"><i class="fa fa-check"></i><b>1.4.1</b> PCA</a></li>
<li class="chapter" data-level="1.4.2" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#variation"><i class="fa fa-check"></i><b>1.4.2</b> Variation</a></li>
</ul></li>
<li class="chapter" data-level="1.5" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#dimension-reduction---hvgs"><i class="fa fa-check"></i><b>1.5</b> Dimension reduction - HVGs</a><ul>
<li class="chapter" data-level="1.5.1" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#pca-using-hvgs"><i class="fa fa-check"></i><b>1.5.1</b> PCA using HVGS</a></li>
<li class="chapter" data-level="1.5.2" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#tsne-does-not-identify-clear-clusters"><i class="fa fa-check"></i><b>1.5.2</b> tSNE does not identify clear clusters</a></li>
<li class="chapter" data-level="1.5.3" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#umap-does-not-identify-clear-clusters"><i class="fa fa-check"></i><b>1.5.3</b> UMAP does not identify clear clusters</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html"><i class="fa fa-check"></i><b>2</b> scBS-seq Data</a><ul>
<li class="chapter" data-level="2.1" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html#data-1"><i class="fa fa-check"></i><b>2.1</b> Data</a></li>
<li class="chapter" data-level="2.2" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html#create-sce"><i class="fa fa-check"></i><b>2.2</b> Create sce</a></li>
<li class="chapter" data-level="2.3" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html#summary"><i class="fa fa-check"></i><b>2.3</b> Summary</a></li>
<li class="chapter" data-level="2.4" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html#variance"><i class="fa fa-check"></i><b>2.4</b> Variance</a></li>
<li class="chapter" data-level="2.5" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html#dimension-reduction"><i class="fa fa-check"></i><b>2.5</b> Dimension reduction</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html"><i class="fa fa-check"></i><b>3</b> scM&amp;T-seq analysis</a><ul>
<li class="chapter" data-level="3.1" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html#correlation"><i class="fa fa-check"></i><b>3.1</b> Correlation</a><ul>
<li class="chapter" data-level="3.1.1" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html#pearson-correlation-study-of-methylation-rate-and-expression---gene-wise"><i class="fa fa-check"></i><b>3.1.1</b> Pearson correlation study of methylation rate and expression - gene wise</a></li>
<li class="chapter" data-level="3.1.2" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html#does-the-variation-in-methylation-in-genebody-correlates-with-variation-in-expression"><i class="fa fa-check"></i><b>3.1.2</b> Does the variation in methylation in genebody correlates with variation in expression?</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html#canonical-correlation-study"><i class="fa fa-check"></i><b>3.2</b> Canonical correlation study</a><ul>
<li class="chapter" data-level="3.2.1" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html#trascriptome---promoter-methylome"><i class="fa fa-check"></i><b>3.2.1</b> Trascriptome - Promoter Methylome</a></li>
<li class="chapter" data-level="3.2.2" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html#run-using-highly-variable-genes"><i class="fa fa-check"></i><b>3.2.2</b> Run using highly variable genes</a></li>
<li class="chapter" data-level="3.2.3" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html#more-strict-filtering-by-coverage"><i class="fa fa-check"></i><b>3.2.3</b> More strict filtering by coverage</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html#outlier-cells-diagnostics"><i class="fa fa-check"></i><b>3.3</b> Outlier cells diagnostics:</a></li>
<li class="chapter" data-level="3.4" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html#block-spls"><i class="fa fa-check"></i><b>3.4</b> Block sPLS</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/AJABADI/scMTseq-AML/" target="blank">See on GitHub</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">scMTseq of AML cells - pilot</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="scmt-seq-analysis" class="section level1">
<h1><span class="header-section-number">Chapter 3</span> scM&amp;T-seq analysis</h1>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">load</span>(<span class="st">&#39;RData/rna_qc.RData&#39;</span>)
<span class="kw">load</span>(<span class="st">&#39;RData/met_sces.RData&#39;</span>)</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## find the common cells in both studies</span>
cell.intersect =<span class="st"> </span><span class="kw">Reduce</span>(intersect, <span class="kw">list</span>(<span class="kw">colnames</span>(sce_genebody), <span class="kw">colnames</span>(rna_qc)))
<span class="kw">length</span>(cell.intersect)

<span class="co">## keep only matching cells</span>
rna_match =<span class="st"> </span>rna_qc[,<span class="kw">colnames</span>(rna_qc) <span class="op">%in%</span><span class="st"> </span>cell.intersect]</code></pre>
<div id="correlation" class="section level2">
<h2><span class="header-section-number">3.1</span> Correlation</h2>
<pre class="sourceCode r"><code class="sourceCode r">pearson_runtime =<span class="st"> </span><span class="kw">system.time</span>({ <span class="co">## 90 sec</span>
  met_sces =<span class="st"> </span><span class="kw">grep</span>(<span class="st">&quot;sce_&quot;</span>, <span class="kw">ls</span>(), <span class="dt">value=</span><span class="ot">TRUE</span>)
  <span class="cf">for</span> (i <span class="cf">in</span> met_sces){
    <span class="kw">assign</span>(i, <span class="kw">corr_mt</span>(rna_match, <span class="kw">get</span>(i), <span class="dt">top_prop =</span> <span class="fl">0.3</span>) )
  }
})
pearson_runtime <span class="op">%&lt;&gt;%</span><span class="st"> </span>.[<span class="dv">1</span>] <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">unname</span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">round</span>()
<span class="kw">save</span>(<span class="dt">list =</span> <span class="kw">c</span>(met_sces, <span class="st">&quot;pearson_runtime&quot;</span>), <span class="dt">file =</span> <span class="st">&quot;RData/met_sces.RData&quot;</span>)</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">met_sces =<span class="st"> </span><span class="kw">grep</span>(<span class="st">&quot;sce_&quot;</span>, <span class="kw">ls</span>(), <span class="dt">value=</span><span class="ot">TRUE</span>)
<span class="kw">load</span>(<span class="st">&quot;RData/met_sces.RData&quot;</span>)</code></pre>
<div id="pearson-correlation-study-of-methylation-rate-and-expression---gene-wise" class="section level3">
<h3><span class="header-section-number">3.1.1</span> Pearson correlation study of methylation rate and expression - gene wise</h3>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## function to create boxplot of pearson correlations of M&amp;T</span>
<span class="co">## for all sces</span>
boxPlot =<span class="st"> </span><span class="cf">function</span>(<span class="dt">sces =</span> met_sces, <span class="dt">n_top =</span> <span class="dv">2000</span>){
  factor_sces =<span class="st"> </span><span class="ot">NULL</span>
  corr =<span class="st"> </span><span class="ot">NULL</span>
  <span class="cf">for</span> (i <span class="cf">in</span> sces){
    
    n =<span class="st"> </span><span class="kw">ifelse</span>(<span class="kw">is.null</span>(n_top),
               <span class="kw">dim</span>(<span class="kw">get</span>(i))[<span class="dv">1</span>], 
               <span class="kw">min</span>(n_top, <span class="kw">dim</span>(<span class="kw">get</span>(i))[<span class="dv">1</span>]))
    
    factor_sces =<span class="st"> </span><span class="kw">c</span>(factor_sces, <span class="kw">rep</span>(i, n))
    corr =<span class="st"> </span><span class="kw">c</span>(corr,<span class="kw">rowData</span>(<span class="kw">get</span>(i)[<span class="dv">1</span><span class="op">:</span>n,])<span class="op">$</span>corr_rna)
  }
  
  p =<span class="st"> </span><span class="kw">ggplot</span>(<span class="dt">data =</span> <span class="ot">NULL</span>, <span class="kw">aes</span>(<span class="dt">x =</span> factor_sces , <span class="dt">y=</span>corr, <span class="dt">fill =</span> <span class="kw">factor</span>(factor_sces))) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">geom_boxplot</span>(<span class="dt">na.rm =</span> <span class="ot">TRUE</span>) <span class="op">+</span><span class="st"> </span><span class="kw">coord_flip</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Pearson correlation&quot;</span>, 
         <span class="dt">title =</span> <span class="st">&quot;Gene-wise correlation of methylation and transcription&quot;</span>, <span class="dt">subtitle =</span> <span class="kw">TeX</span>(<span class="kw">paste0</span>(<span class="st">&quot;top &quot;</span>,n_top, <span class="st">&quot; genes based on &quot;</span>,<span class="st">&quot;$</span><span class="ch">\\</span><span class="st">hat{</span><span class="ch">\\</span><span class="st">nu}$&quot;</span>))) <span class="op">+</span>
<span class="st">    </span><span class="kw">guides</span>(<span class="dt">fill=</span><span class="ot">FALSE</span>)
  <span class="kw">return</span>(p)
  }</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## Pearson correlation boxplots</span>
<span class="kw">boxPlot</span>()</code></pre>
<p><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-52-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="does-the-variation-in-methylation-in-genebody-correlates-with-variation-in-expression" class="section level3">
<h3><span class="header-section-number">3.1.2</span> Does the variation in methylation in genebody correlates with variation in expression?</h3>
<pre class="sourceCode r"><code class="sourceCode r">var_corr =<span class="st"> </span><span class="kw">data.frame</span>(<span class="kw">matrix</span>(<span class="dt">nrow =</span> <span class="kw">length</span>(met_sces),<span class="dt">ncol =</span> <span class="dv">2</span>),
                      <span class="dt">row.names =</span> met_sces )
<span class="kw">colnames</span>(var_corr) =<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;total&quot;</span>, <span class="st">&quot;biological&quot;</span>)
<span class="cf">for</span> (i <span class="cf">in</span> met_sces){
  var_corr[i,] =<span class="st"> </span><span class="kw">c</span>(<span class="kw">metadata</span>(<span class="kw">get</span>(i))<span class="op">$</span>corr_var_tot_wtd_by_cov[<span class="dv">1</span>],
                   <span class="kw">metadata</span>(<span class="kw">get</span>(i))<span class="op">$</span>corr_var_bio_wtd_by_cov[<span class="dv">1</span>])
}

<span class="kw">ggplot</span>() <span class="op">+</span><span class="st"> </span><span class="kw">geom_col</span>(<span class="kw">aes</span>(<span class="kw">factor</span>(<span class="kw">rownames</span>(var_corr)),var_corr<span class="op">$</span>total, <span class="dt">fill =</span> <span class="kw">factor</span>(<span class="kw">rownames</span>(var_corr))), <span class="dt">shape=</span><span class="dv">23</span>, <span class="dt">size=</span><span class="dv">4</span>) <span class="op">+</span><span class="st"> </span><span class="kw">guides</span>(<span class="dt">fill=</span><span class="ot">FALSE</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Genomic Context&quot;</span>, <span class="dt">y=</span> <span class="st">&quot;Pearson correlation&quot;</span>, <span class="dt">title =</span> <span class="st">&quot;Correlation of total transcrption variance </span><span class="ch">\n</span><span class="st"> and weighted methylation variance&quot;</span>) <span class="op">+</span><span class="st"> </span><span class="kw">coord_cartesian</span>(<span class="dt">ylim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="fl">0.3</span>, <span class="fl">0.3</span>))</code></pre>
<p><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-53-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>() <span class="op">+</span><span class="st"> </span><span class="kw">geom_col</span>(<span class="kw">aes</span>(<span class="kw">factor</span>(<span class="kw">rownames</span>(var_corr)),var_corr<span class="op">$</span>bio, <span class="dt">fill =</span> <span class="kw">factor</span>(<span class="kw">rownames</span>(var_corr))), <span class="dt">shape=</span><span class="dv">23</span>, <span class="dt">size=</span><span class="dv">4</span>) <span class="op">+</span><span class="st"> </span><span class="kw">guides</span>(<span class="dt">fill=</span><span class="ot">FALSE</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Genomic Context&quot;</span>, <span class="dt">y=</span> <span class="st">&quot;Pearson correlation&quot;</span>, <span class="dt">title =</span> <span class="st">&quot;Correlation of decomposed biological transcrption variance </span><span class="ch">\n</span><span class="st"> and weighted methylation variance&quot;</span>)<span class="op">+</span><span class="st"> </span><span class="kw">coord_cartesian</span>(<span class="dt">ylim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="fl">0.3</span>, <span class="fl">0.3</span>))</code></pre>
<p><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-53-2.png" width="672" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="canonical-correlation-study" class="section level2">
<h2><span class="header-section-number">3.2</span> Canonical correlation study</h2>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## dataset parameters for plot</span>
<span class="co">## first one will be taken as Y, the rest X&#39;s</span>
<span class="co">## all that are in the sPLS list + RNA</span>
par_plot =<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">row.names =</span> <span class="kw">c</span>(<span class="st">&quot;rna_match&quot;</span>, <span class="st">&quot;sce_genebody&quot;</span>, <span class="st">&quot;sce_prom&quot;</span>, <span class="st">&quot;sce_prom_cgi&quot;</span>, <span class="st">&quot;sce_prom_noncgi&quot;</span> ),
                      <span class="dt">label =</span> <span class="kw">c</span>(<span class="st">&quot;Transcriptome&quot;</span>,<span class="st">&quot;Genebody&quot;</span>, <span class="st">&quot;Promoter&quot;</span>, <span class="st">&quot;Promoter (CGI)&quot;</span>, <span class="st">&quot;Promoter (non-CGI)&quot;</span>),
                      <span class="dt">shape =</span><span class="kw">c</span>(<span class="dv">17</span>, <span class="dv">19</span>, <span class="dv">15</span>, <span class="dv">18</span>, <span class="dv">8</span>), <span class="co">## shapes for corr circ plots</span>
                      <span class="dt">col =</span> <span class="kw">color.mixo</span>(<span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">4</span>, <span class="dv">5</span>, <span class="dv">8</span>))) <span class="co">## colors for corr circ plots</span></code></pre>
<div id="trascriptome---promoter-methylome" class="section level3">
<h3><span class="header-section-number">3.2.1</span> Trascriptome - Promoter Methylome</h3>
<!--PICKUP-->
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## spls for all methylation data against transcriptome</span>
spls_all_runtime =<span class="st"> </span><span class="kw">system.time</span>({ <span class="co">## 64 sec</span>
  spls_all =<span class="st"> </span><span class="kw">spls_wrapper</span>(<span class="dt">x=</span><span class="dv">2</span><span class="op">:</span><span class="dv">5</span>, <span class="dt">block =</span> <span class="ot">FALSE</span>)
})
spls_runtime <span class="op">%&lt;&gt;%</span><span class="st"> </span>.[<span class="dv">1</span>] <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">unname</span>()
<span class="kw">save</span>(spls_all,spls_all_runtime, <span class="dt">file=</span><span class="st">&quot;RData/spls_all.RData&quot;</span> )</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">source</span>(<span class="st">&quot;utils/spls_plot.R&quot;</span>)
<span class="kw">spls_plot</span>(spls_all)</code></pre>
<p><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-57-1.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-57-2.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-57-3.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-57-4.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-57-5.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-57-6.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-57-7.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-57-8.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-57-9.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-57-10.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-57-11.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-57-12.png" width="672" style="display: block; margin: auto;" /></p>
<p>H8 was also an outlier from PCA analysis. So we will exclude it.</p>
<pre class="sourceCode r"><code class="sourceCode r">outliers =<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;H8&quot;</span>)
spls_all_ex_H8 =<span class="st"> </span><span class="kw">spls_wrapper</span>(<span class="dt">x=</span><span class="dv">2</span><span class="op">:</span><span class="dv">5</span>, <span class="dt">block =</span> <span class="ot">FALSE</span>, <span class="dt">exclude =</span> outliers)
<span class="kw">save</span>(spls_all_ex_H8, <span class="dt">file=</span><span class="st">&quot;RData/spls_all_ex_H8.RData&quot;</span> )</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">source</span>(<span class="st">&quot;utils/spls_plot.R&quot;</span>)
<span class="kw">spls_plot</span>(spls_all_ex_H8)</code></pre>
<p><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-60-1.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-60-2.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-60-3.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-60-4.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-60-5.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-60-6.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-60-7.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-60-8.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-60-9.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-60-10.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-60-11.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-60-12.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="run-using-highly-variable-genes" class="section level3">
<h3><span class="header-section-number">3.2.2</span> Run using highly variable genes</h3>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## spls for all methylation data against transcriptome - top 30% variable genes</span>
spls_ex_H8_top_30pct =<span class="st"> </span><span class="kw">spls_wrapper</span>(<span class="dt">x=</span><span class="dv">2</span><span class="op">:</span><span class="dv">5</span>, <span class="dt">block =</span> <span class="ot">FALSE</span>, <span class="dt">top_x =</span> <span class="fl">0.3</span>, <span class="dt">top_y =</span> <span class="fl">0.3</span>, <span class="dt">exclude =</span> outliers)
<span class="kw">save</span>(spls_ex_H8_top_30pct, <span class="dt">file=</span><span class="st">&quot;RData/spls_ex_H8_top_30pct.RData&quot;</span> )</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">spls_plot</span>(spls_ex_H8_top_30pct)</code></pre>
<p><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-63-1.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-63-2.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-63-3.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-63-4.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-63-5.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-63-6.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-63-7.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-63-8.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-63-9.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-63-10.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-63-11.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-63-12.png" width="672" style="display: block; margin: auto;" /></p>
<p>Less outlier problem and more agreement problem.</p>
</div>
<div id="more-strict-filtering-by-coverage" class="section level3">
<h3><span class="header-section-number">3.2.3</span> More strict filtering by coverage</h3>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## spls for all methylation data against transcriptome -  coverage &gt;=75%</span>
spls_ex_H8_cov_<span class="fl">0.75</span> =<span class="st"> </span><span class="kw">spls_wrapper</span>(<span class="dt">x=</span><span class="dv">2</span><span class="op">:</span><span class="dv">5</span>, <span class="dt">block =</span> <span class="ot">FALSE</span>, <span class="dt">cov =</span> <span class="fl">0.75</span>, <span class="dt">exclude =</span> outliers)
<span class="kw">save</span>(spls_ex_H8_cov_<span class="fl">0.75</span>, <span class="dt">file=</span><span class="st">&quot;RData/spls_ex_H8_cov_0.75.RData&quot;</span> )
<span class="co"># Number of genes remained in methylation: 11525</span>
<span class="co"># Number of genes remained in methylation: 575</span>
<span class="co"># Number of genes remained in methylation: 401</span>
<span class="co"># Number of genes remained in methylation: 174</span></code></pre>
<p>The number of genes that pased filtering was low in promoter methylation datasets. D5 and D7 show outlier behaviour.
Sample and Arrow plots without the outlier cells:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">spls_plot</span>(spls_ex_H8_cov_<span class="fl">0.75</span>, <span class="dt">hide =</span> <span class="kw">c</span>(<span class="st">&quot;D5&quot;</span>, <span class="st">&quot;D7&quot;</span>), <span class="dt">plots =</span> <span class="kw">c</span>(<span class="st">&quot;plotIndiv&quot;</span>, <span class="st">&quot;plotArrow&quot;</span>))</code></pre>
<p><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-66-1.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-66-2.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-66-3.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-66-4.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-66-5.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-66-6.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-66-7.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-66-8.png" width="672" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="outlier-cells-diagnostics" class="section level2">
<h2><span class="header-section-number">3.3</span> Outlier cells diagnostics:</h2>
<p>Total number of missing values for all cells:</p>
<pre class="sourceCode r"><code class="sourceCode r">outliers =<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;H8&quot;</span>, <span class="st">&quot;H6&quot;</span>, <span class="st">&quot;H7&quot;</span>, <span class="st">&quot;D5&quot;</span>, <span class="st">&quot;D7&quot;</span>, <span class="st">&quot;E8&quot;</span>)
NAs =<span class="st"> </span><span class="kw">colSums</span>(<span class="kw">is.na</span>(<span class="kw">assay</span>(sce_genebody,<span class="st">&quot;rates&quot;</span>)))
cols =<span class="st"> </span><span class="kw">rep</span>(<span class="st">&quot;darkblue&quot;</span>, <span class="kw">length</span>(NAs))
<span class="kw">names</span>(cols) =<span class="st"> </span><span class="kw">names</span>(NAs)
cols[outliers] =<span class="st"> &quot;red&quot;</span>
<span class="kw">ggplot</span>() <span class="op">+</span><span class="st"> </span><span class="kw">geom_text</span>(<span class="kw">aes</span>(NAs, <span class="kw">names</span>(NAs), <span class="dt">label =</span> <span class="kw">names</span>(NAs)), <span class="dt">col =</span> cols) <span class="op">+</span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">x=</span> <span class="st">&quot;Total number of NAs&quot;</span>, <span class="dt">y=</span><span class="st">&quot;Cell&quot;</span>)</code></pre>
<p><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-67-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>From outlier cells, only H8 has relatively high number of NA’s. Three of outliers are adjacent cells, which indicates there might have been technical effect.</p>
<p>Total number of observations all cells:</p>
<pre class="sourceCode r"><code class="sourceCode r">weight_sums =<span class="st"> </span><span class="kw">colSums</span>(<span class="kw">assay</span>(sce_genebody, <span class="st">&quot;weights&quot;</span>), <span class="dt">na.rm =</span> <span class="ot">TRUE</span>)
cols =<span class="st"> </span><span class="kw">rep</span>(<span class="st">&quot;darkblue&quot;</span>, <span class="kw">length</span>(NAs))
<span class="kw">names</span>(cols) =<span class="st"> </span><span class="kw">names</span>(weight_sums)
cols[outliers] =<span class="st"> &quot;red&quot;</span>
<span class="kw">ggplot</span>() <span class="op">+</span><span class="st"> </span><span class="kw">geom_text</span>(<span class="kw">aes</span>(weight_sums, <span class="kw">names</span>(weight_sums), <span class="dt">label =</span> <span class="kw">names</span>(weight_sums)), <span class="dt">col =</span> cols) <span class="op">+</span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">x=</span> <span class="st">&quot;Total weights for cells&quot;</span>, <span class="dt">y=</span><span class="st">&quot;Cell&quot;</span>)</code></pre>
<p><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-68-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Not a clear cause.</p>
</div>
<div id="block-spls" class="section level2">
<h2><span class="header-section-number">3.4</span> Block sPLS</h2>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">source</span>(<span class="st">&quot;utils/block_spls_plot.R&quot;</span>)
<span class="co">## Y: transcriptome, X = c(genebody, promoter) methylome</span>
block_gbody_prom =<span class="st"> </span><span class="kw">spls_wrapper</span>(<span class="dt">df =</span> par_plot, <span class="dt">x =</span> <span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">3</span>),<span class="dt">y =</span> <span class="dv">1</span>, <span class="dt">ncomp =</span><span class="dv">2</span>,
                          <span class="dt">top_x =</span> <span class="ot">NULL</span>, <span class="dt">top_y =</span> <span class="ot">NULL</span>, <span class="dt">cov =</span> <span class="ot">NULL</span>, <span class="dt">exclude=</span><span class="st">&quot;H8&quot;</span>,
                          <span class="dt">keepX=</span><span class="kw">c</span>(<span class="dv">50</span>,<span class="dv">50</span>), <span class="dt">keepY=</span><span class="kw">c</span>(<span class="dv">50</span>,<span class="dv">50</span>), <span class="dt">block=</span><span class="ot">TRUE</span>, <span class="dt">mode=</span><span class="st">&quot;canonical&quot;</span>)
<span class="kw">save</span>(block_gbody_prom, <span class="dt">file=</span><span class="st">&quot;RData/block_gbody_prom.RData&quot;</span>)</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">load</span>(<span class="st">&quot;RData/block_gbody_prom.RData&quot;</span>)</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">block_spls_plot</span>(<span class="dt">block =</span> block_gbody_prom, <span class="dt">pars =</span> par_plot, <span class="dt">x=</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">3</span>), <span class="dt">y=</span><span class="dv">1</span>)</code></pre>
<p><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-71-1.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-71-2.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-71-3.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## Y: transcriptome, X = c(genebody, promoter CGI) methylome</span>
block_gbody_prom_cgi =<span class="st"> </span><span class="kw">spls_wrapper</span>(<span class="dt">df =</span> par_plot, <span class="dt">x =</span> <span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">4</span>),<span class="dt">y =</span> <span class="dv">1</span>, <span class="dt">ncomp =</span><span class="dv">2</span>,
                          <span class="dt">top_x =</span> <span class="ot">NULL</span>, <span class="dt">top_y =</span> <span class="ot">NULL</span>, <span class="dt">cov =</span> <span class="ot">NULL</span>, <span class="dt">exclude=</span><span class="st">&quot;H8&quot;</span>,
                          <span class="dt">keepX=</span><span class="kw">c</span>(<span class="dv">50</span>,<span class="dv">50</span>), <span class="dt">keepY=</span><span class="kw">c</span>(<span class="dv">50</span>,<span class="dv">50</span>), <span class="dt">block=</span><span class="ot">TRUE</span>, <span class="dt">mode=</span><span class="st">&quot;canonical&quot;</span>)
<span class="kw">save</span>(block_gbody_prom_cgi, <span class="dt">file=</span><span class="st">&quot;RData/block_gbody_prom_cgi.RData&quot;</span>)</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">block_spls_plot</span>(<span class="dt">block =</span> block_gbody_prom_cgi, <span class="dt">pars =</span> par_plot, <span class="dt">x=</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">4</span>), <span class="dt">y=</span><span class="dv">1</span>)</code></pre>
<p><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-73-1.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-73-2.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-73-3.png" width="672" style="display: block; margin: auto;" />
E8 and H6 consistently showed outlier behaviour. So we’ll filter them and repeat the analysis.</p>
<pre class="sourceCode r"><code class="sourceCode r">outliers =<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;H8&quot;</span>, <span class="st">&quot;H6&quot;</span>, <span class="st">&quot;E8&quot;</span>)</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">source</span>(<span class="st">&quot;utils/block_spls_plot.R&quot;</span>)
<span class="co">## Y: transcriptome, X = c(genebody, promoter) methylome</span>
block_gbody_prom_outliers =<span class="st"> </span><span class="kw">spls_wrapper</span>(<span class="dt">df =</span> par_plot, <span class="dt">x =</span> <span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">3</span>),<span class="dt">y =</span> <span class="dv">1</span>, <span class="dt">ncomp =</span><span class="dv">2</span>,
                          <span class="dt">top_x =</span> <span class="ot">NULL</span>, <span class="dt">top_y =</span> <span class="ot">NULL</span>, <span class="dt">cov =</span> <span class="ot">NULL</span>, <span class="dt">exclude=</span>outliers,
                          <span class="dt">keepX=</span><span class="kw">c</span>(<span class="dv">50</span>,<span class="dv">50</span>), <span class="dt">keepY=</span><span class="kw">c</span>(<span class="dv">50</span>,<span class="dv">50</span>), <span class="dt">block=</span><span class="ot">TRUE</span>, <span class="dt">mode=</span><span class="st">&quot;canonical&quot;</span>)
<span class="kw">save</span>(block_gbody_prom_outliers, <span class="dt">file=</span><span class="st">&quot;RData/block_gbody_prom_outliers.RData&quot;</span>)</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">load</span>(<span class="st">&quot;RData/block_gbody_prom_outliers.RData&quot;</span>)</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">block_spls_plot</span>(<span class="dt">block =</span> block_gbody_prom_outliers, <span class="dt">pars =</span> par_plot, <span class="dt">x=</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">3</span>), <span class="dt">y=</span><span class="dv">1</span>)</code></pre>
<p><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-77-1.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-77-2.png" width="672" style="display: block; margin: auto;" /><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-77-3.png" width="672" style="display: block; margin: auto;" /></p>
<!-- ```{r} -->
<!-- plotVar(block.spls, legend = T, title = 'Correlation Circle Plots - Transcriptome against Methylation', pch = c(16,20,17), col= color.mixo(c(2,4,1))) -->
<!-- ## see if there are outliers in the data using sample and arrow plots -->
<!-- plotIndiv(block.spls) -->
<!-- plotArrow(block.spls, ind.names = T) -->
<!-- ``` -->
<!-- ```{r} -->
<!-- ## all genes in latent factors -->
<!-- genes.rna = c( -->
<!--   selectVar(block.spls, comp = 1)$Y$name, -->
<!--   selectVar(block.spls, comp = 2)$Y$name -->
<!-- ) -->
<!-- genes.body = c( -->
<!--   selectVar(block.spls, comp = 1)$gene.met$name, -->
<!--   selectVar(block.spls, comp = 2)$gene.met$name -->
<!-- ) -->
<!-- genes.prom = c( -->
<!--   selectVar(block.spls, comp = 1)$prom.met$name, -->
<!--   selectVar(block.spls, comp = 2)$prom.met$name -->
<!--   ) -->
<!-- genes = c(genes.rna, genes.body, genes.prom) -->
<!-- length(genes) -->
<!-- ## only 3 repetitions -->
<!-- length(unique(genes)) -->
<!-- ## duplicated genes -->
<!-- genes[duplicated(genes)] -->
<!-- ## in which dataset are they? -->
<!-- which(genes == genes[duplicated(genes)][1]) -->
<!-- ## comp1 genebody and comp 1 prom -->
<!-- which(genes == genes[duplicated(genes)][2]) -->
<!-- ## comp1 rna and comp 2 prom -->
<!-- which(genes == genes[duplicated(genes)][3]) -->
<!-- ## comp2 rna and comp 2 prom -->
<!-- ``` -->
<!-- #### Do the signature genes exhibit different correlation structures? -->
<!-- ```{r} -->
<!-- boxplot(genebody.cor$pearson[genes.body] , horizontal=T) -->
<!-- boxplot(prom.cor$pearson[genes.prom] , horizontal=T) -->
<!-- ``` -->
<!-- ```{r, warning=F, message =F} -->
<!-- ggplot(genebody.cor$vars$all[genes.body,]) + geom_point(aes(x= met, y = expr), col = 'darkblue') + -->
<!--   labs(x= "variation in genebody methylation", y="variation in gene expression", title = 'correlation b/w variation in expression and genebody methylation - signature genes') -->
<!-- ``` -->
<!-- ```{r, warning=F, message=F} -->
<!-- ggplot(prom.cor$vars$all[genes.prom,]) + geom_point(aes(x= met, y = expr), col = 'darkblue') + -->
<!--   labs(x= "variation in promoter methylation", y="variation in gene expression", title = 'correlation b/w variation in expression and promoter methylation - signature genes') -->
<!-- ``` -->
<!-- #### Q: are the signature genes biologicall dependent? -->
<!-- ## 3kb Windows -->
<!-- <!-- -->
<!-- Add the option to re-summarise the methylation data here: -->
<!-- if(!io$use.parsed){ -->
<!-- } else { -->
<!-- } -->
<!-- -->
<p>–&gt;
<!-- ```{r, eval=F} -->
<!-- system.time({ -->
<!--  ##   user  system elapsed -->
<!--  ## 18.916   5.696  36.746 -->
<!-- window3kb = fread(io$met_3kb) -->
<!-- }) -->
<!-- ``` -->
<!-- ```{r, echo=F} -->
<!-- # save(window3kb, file='window3kb.RData') -->
<!-- load('window3kb.RData') -->
<!-- ``` -->
<!-- ```{r} -->
<!-- ## number of seqnames -->
<!-- dim(window3kb)[1] ## 1,923,673 -->
<!-- ## valid chromosomes -->
<!-- valid.chr = c(as.character(1:22), 'MT', 'X', 'Y') -->
<!-- ## some invalid chromoses from data -->
<!-- head(window3kb$seqnames[!window3kb$seqnames %in% valid.chr]) -->
<!-- ## number of invalid seqnames -->
<!-- length(window3kb$seqnames[!window3kb$seqnames %in% valid.chr]) ## 5650 -->
<!-- ## filter seqnames to keep valid chromosomes only -->
<!-- window3kb = window3kb[seqnames %in% valid.chr] --></p>
<!-- ``` -->
<!-- ```{r} -->
<!-- # window3kb.copy = window3kb -->
<!-- window3kb.overlap = window3kb -->
<!-- system.time({ -->
<!--           ##  user  system elapsed -->
<!--           ## 4.073   2.501   7.482 -->
<!-- ## create a 'bin' variable -->
<!-- ## using 1e5 bins made it too uniform, i.e. meth. rate ~0.75 -->
<!-- ## sticking with the 3kb window -->
<!-- bin.width  =3e3 -->
<!-- ## I have not differentiated between forward and backward strand, to be checked with Heather -->
<!-- ## [Al] also, some windows overlap - this should be considered later on (weighted ovrlapping) -->
<!-- window3kb[,bin:=ceiling(start/bin.width)] -->
<!-- ## dropping all columns except for rates and bin -->
<!-- window3kb = window3kb[, grepl('rate', colnames(window3kb))| -->
<!--                         grepl('bin', colnames(window3kb)), with=F] -->
<!-- ## dropping all columns except for rates and start -->
<!-- window3kb.overlap = window3kb.overlap[, grepl('rate', colnames(window3kb.overlap))| -->
<!--                         grepl('start', colnames(window3kb.overlap)), with=F] -->
<!-- ## remove every column name that matches 'var' - rate variance data -->
<!-- window3kb = window3kb[,!grepl('var', colnames(window3kb)), with=F] -->
<!-- window3kb.overlap = window3kb.overlap[,!grepl('var', colnames(window3kb.overlap)), with=F] -->
<!-- dim(window3kb) # 1923673      41 -->
<!-- dim(window3kb.overlap) -->
<!-- source('utils/name.cells.by.wells.R') -->
<!-- colnames(window3kb) = name.cells.by.wells(colnames(window3kb), pos.in.string = 'first') -->
<!-- colnames(window3kb.overlap) = name.cells.by.wells(colnames(window3kb.overlap), pos.in.string = 'first') -->
<!-- ## create a 'weight' variable for the number of observations in each bin -->
<!-- ## [Al] although, what also matters is the number of times each cell has been covered in total -->
<!-- window3kb = window3kb[,weight:=.N, by = bin] -->
<!-- window3kb.overlap = window3kb.overlap[,s.weight:=.N, by = start] -->
<!--  #   user  system elapsed -->
<!--  # 22.470  15.497  54.818 -->
<!-- ## average over bins and disregard NAs- weights will remain the same so not excluded -->
<!-- ## bins with no observation for a given cell will return NaNs -->
<!-- window3kb = window3kb[,lapply(.SD,mean, na.rm=TRUE), by=bin] -->
<!-- window3kb.overlap = window3kb.overlap[,lapply(.SD,mean, na.rm=TRUE), by=start] -->
<!-- dim(window3kb) ## 82981    42 -->
<!-- window3kb$n.NAs = rowSums(is.na(window3kb)) -->
<!-- window3kb.overlap$n.NAs = rowSums(is.na(window3kb.overlap)) -->
<!-- }) -->
<!-- ``` -->
<!-- ```{r} -->
<!-- hist(window3kb$n.NAs, main='histogram of number of NAs in binned 3kb windows') -->
<!-- hist(window3kb.overlap$n.NAs, main='histogram of number of NAs in all 3kb windows') -->
<!-- ``` -->
<!-- ```{r, eval=F} -->
<!-- ## filter out NAs -->
<!-- window3kb = window3kb[n.NAs<=10] ## arbitrary -->
<!-- window3kb.overlap = window3kb.overlap[n.NAs<=10] ## arbitrary -->
<!-- dim(window3kb) ## 74k    43 -->
<!-- dim(window3kb.overlap) ## 138k    43 -->
<!-- ``` -->
<!-- ```{r} -->
<!-- ## plot a given cell's weight -->
<!-- plot(window3kb$weight, as.matrix(window3kb[,9]), xlim = c(0,250), pch=16, col=rgb(red=0.1, green=0.2, blue=0.7, alpha=0.3), -->
<!--      main= 'methylation rate at each binned window for a \n given cell vs the window coverage') -->
<!-- plot(window3kb.overlap$s.weight, as.matrix(window3kb.overlap[,9]), xlim = c(0,250), pch=16, col=rgb(red=0.1, green=0.2, blue=0.7, alpha=0.3), -->
<!--      main= 'methylation rate at each window for a \n given cell vs the window coverage') -->
<!-- ## part of variation in methylation is due to dropouts, i.e. low weight -->
<!-- ## plot the boxplot of weights -->
<!-- boxplot(window3kb$weight, horizontal=T) -->
<!-- ``` -->
<!-- ```{r, eval=F} -->
<!-- ## filtering by weight -->
<!-- ## require each window to be covered at least x times -->
<!-- min.weight = 20 ## arbitrary -->
<!-- window3kb = window3kb[weight > min.weight] -->
<!-- dim(window3kb) ##  35431    43  - for min.weight = median -->
<!-- ``` -->
<!-- ```{r, eval=F} -->
<!-- system.time({ -->
<!--     ##  user  system elapsed -->
<!--     ## 24.462   0.207  24.994 -->
<!-- # ## add mean meth. rate for each bin: -->
<!-- window3kb[,mean:=apply(.SD,1, mean, na.rm=TRUE ) , by=bin, .SDcols=c(2:41)] -->
<!-- window3kb.overlap[,mean:=apply(.SD,1, mean, na.rm=TRUE ) , by=start, .SDcols=c(2:41)] -->
<!-- ## checking whether there are near zero means for calculating cv -->
<!-- hist(window3kb$mean, main='mean methylation rate for all cells - binned 3kb windows') -->
<!-- hist(window3kb.overlap$mean, main='mean methylation rate for all cells - 3kb windows') -->
<!-- min(window3kb$mean)## 0.5275498 -->
<!-- ## so we'll calculate cv with no pseudo value at denominator -->
<!-- window3kb[,sd:=apply(.SD,1, sd, na.rm=TRUE) , by=bin,.SDcols=c(2:41)] -->
<!-- window3kb[,cv:=sd/mean] -->
<!-- window3kb.overlap[,sd:=apply(.SD,1, sd, na.rm=TRUE) , by=start,.SDcols=c(2:41)] -->
<!-- window3kb.overlap[,cv:=sd/mean] -->
<!-- plot(window3kb$cv~window3kb$mean, col = 'darkblue', main = 'BCV for 3kb windows') -->
<!-- ## order by cv for highly varibale regions -->
<!-- window3kb=window3kb[order(-cv)] -->
<!-- window3kb.overlap=window3kb.overlap[order(-cv)] -->
<!-- ## keep the top variable windows -->
<!-- n.region = 5000 -->
<!-- window3kb.hvw=window3kb[1:n.region] -->
<!-- window3kb.overlap.hvw=window3kb.overlap[1:n.region] -->
<!-- ## create SCE -->
<!-- hvw.sce = SingleCellExperiment(assays = list(counts = as.matrix(window3kb.hvw[,c(2:41)]))) -->
<!-- hvw.overlap.sce = SingleCellExperiment(assays = list(counts = as.matrix(window3kb.overlap.hvw[,c(2:41)]))) -->
<!-- }) -->
<!-- ``` -->
<!-- ### Dimension Reduction -->
<!-- ```{r} -->
<!-- system.time({ -->
<!--      ## user  system elapsed -->
<!--     ## 253.382  31.798 286.293 -->
<!-- # pca.met.window = pca(t(counts(hvw.sce)), ncomp = 3) -->
<!-- # save(pca.met.window, file= file.path(io$RData,'pca.met.window.RData')) -->
<!-- load(file.path(io$RData,'pca.met.window.RData')) -->
<!-- }) -->
<!-- ## outliers exist: -->
<!-- source('utils/plot.pca.R') -->
<!-- plot.pca(gg.pca = gg.pca, pca.obj = pca.met.window, top = 'PCA - methylation in variable 3kb windows') -->
<!-- ## find outliers -->
<!-- variates =as.data.frame(pca.met.window$variates$X) -->
<!-- variates.ordered = variates[order((abs(variates$PC1)+abs(variates$PC2)+abs(variates$PC3)), decreasing = T),] -->
<!-- ## plot labeled pca to find outliers -->
<!-- ggplot(variates.ordered, aes(x = PC1, y = PC2)) + -->
<!--   geom_point(size = 3, col=color.mixo(1)) + -->
<!--   geom_label_repel( -->
<!--     aes(label = ifelse(rank(PC1)<10,rownames(variates),'')), -->
<!--     col = 'darkorange', -->
<!--     fontface = 'bold', -->
<!--     box.padding = unit(0.35, "lines"), -->
<!--     point.padding = unit(0.5, "lines"), -->
<!--     segment.color = 'grey50' -->
<!--   ) -->
<!-- ## if filtering necessary -->
<!-- # outlier.cells = c('A7','A5','A6', 'A8', 'A9') -->
<!-- ## Were the A row's conditions different? -->
<!-- ``` -->
<!-- ```{r, eval=F} -->
<!-- spls.rna.hvw  = spls(Y=t(logcounts(rna.sce)), X=t(counts(hvw.sce)), ncomp = 2, keepX = c(50,50), keepY= c(50,50), mode = "canonical") -->
<!-- spls.rna.hvw.overlap  = spls(Y=t(logcounts(rna.sce)), X=t(counts(hvw.overlap.sce)), ncomp = 2, keepX = c(50,50), keepY= c(50,50), mode = "canonical") -->
<!-- ``` -->
<!-- ```{r} -->
<!-- plotIndiv(spls.rna.hvw) -->
<!-- plotIndiv(spls.rna.hvw.overlap) -->
<!-- ``` -->
<!-- ```{r} -->
<!-- ## the data points for methylation data represents a binned window -->
<!-- plotVar(spls.rna.hvw, legend = T, title = paste0('Correlation Circle Plots - Transcriptome (Y) against Methylation \n of the binned ',as.integer(bin.width/1000),' kb windows (X)'), pch= c(16,17), col= color.mixo(c(5,1))) -->
<!-- ## the data points for methylation data represents a 3kb window -->
<!-- plotVar(spls.rna.hvw.overlap, legend = T, title = paste0('Correlation Circle Plots - Transcriptome (Y) against Methylation \n of the unbiased ',as.integer(bin.width/1000),' kb windows (X)'), pch= c(16,17), col= color.mixo(c(5,1))) -->
<!-- ``` -->
<!-- Negative correlation between RNAseq and Methylation in 3kb window data. -->
<!-- ```{r, echo=F} -->
<!-- if(io$save.RData){ -->
<!--   save.name = paste0('scMT_met_', Sys.info()['user'],'_',format(Sys.Date(), format='%y%m%d'), '.RData') -->
<!--   save.image(file = file.path(io$RData,save.name)) -->
<!-- } -->
<!-- ``` -->

<div id="refs" class="references">
<div>
<p>Lun, Aaron T.L., Davis J. McCarthy, and John C. Marioni. 2018. “A Step-by-Step Workflow for Low-Level Analysis of Single-Cell Rna-Seq Data with Bioconductor.” <em>F1000Research</em> 5 (2122). <a href="https://doi.org/10.12688/f1000research.9501.2">https://doi.org/10.12688/f1000research.9501.2</a>.</p>
</div>
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="scbs-seq-data.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
