<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Chapter 1 scRNA-seq Data | scMTseq of AML cells - pilot</title>
  <meta name="description" content="Joint methylation and transcriptome sequencing of AML cells">
  <meta name="generator" content="bookdown  and GitBook 2.6.7">

  <meta property="og:title" content="Chapter 1 scRNA-seq Data | scMTseq of AML cells - pilot" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Joint methylation and transcriptome sequencing of AML cells" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 1 scRNA-seq Data | scMTseq of AML cells - pilot" />
  
  <meta name="twitter:description" content="Joint methylation and transcriptome sequencing of AML cells" />
  

<meta name="author" content="Al J Abadi">


<meta name="date" content="2019-02-07">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="index.html">
<link rel="next" href="scbs-seq-data.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />







<script type="text/javascript">

// toggle visibility of R source blocks in R Markdown output
function toggle_R() {
  var x = document.getElementsByClassName('r');
  if (x.length == 0) return;
  function toggle_vis(o) {
    var d = o.style.display;
    o.style.display = (d == 'block' || d == '') ? 'none':'block';
  }

  for (i = 0; i < x.length; i++) {
    var y = x[i];
    if (y.tagName.toLowerCase() === 'pre') toggle_vis(y);
  }

    var elem = document.getElementById("myButton1");
    if (elem.value === "Hide Global") elem.value = "Show Global";
    else elem.value = "Hide Global";
}

document.write('<input onclick="toggle_R();" type="button" value="Hide Global" id="myButton1" style="position: absolute; top: 10%; right: 2%; z-index: 200"></input>')

</script>


<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Pilot scMTseq of AML cells</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Data</a></li>
<li class="chapter" data-level="1" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html"><i class="fa fa-check"></i><b>1</b> scRNA-seq Data</a><ul>
<li class="chapter" data-level="1.1" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#load-raw-data"><i class="fa fa-check"></i><b>1.1</b> Load Raw Data</a></li>
<li class="chapter" data-level="1.2" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#preprocessing"><i class="fa fa-check"></i><b>1.2</b> Preprocessing</a><ul>
<li class="chapter" data-level="1.2.1" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#qc-metrics"><i class="fa fa-check"></i><b>1.2.1</b> QC metrics</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#normalisation"><i class="fa fa-check"></i><b>1.3</b> Normalisation</a><ul>
<li class="chapter" data-level="1.3.1" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#scran"><i class="fa fa-check"></i><b>1.3.1</b> scran</a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#dimension-reduction---filtered-genes"><i class="fa fa-check"></i><b>1.4</b> Dimension reduction - filtered genes</a><ul>
<li class="chapter" data-level="1.4.1" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#pca"><i class="fa fa-check"></i><b>1.4.1</b> PCA</a></li>
<li class="chapter" data-level="1.4.2" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#variation"><i class="fa fa-check"></i><b>1.4.2</b> Variation</a></li>
</ul></li>
<li class="chapter" data-level="1.5" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#dimension-reduction---hvgs"><i class="fa fa-check"></i><b>1.5</b> Dimension reduction - HVGs</a><ul>
<li class="chapter" data-level="1.5.1" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#pca-using-hvgs"><i class="fa fa-check"></i><b>1.5.1</b> PCA using HVGS</a></li>
<li class="chapter" data-level="1.5.2" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#tsne-does-not-identify-clear-clusters"><i class="fa fa-check"></i><b>1.5.2</b> tSNE does not identify clear clusters</a></li>
<li class="chapter" data-level="1.5.3" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#umap-does-not-identify-clear-clusters"><i class="fa fa-check"></i><b>1.5.3</b> UMAP does not identify clear clusters</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html"><i class="fa fa-check"></i><b>2</b> scBS-seq Data</a><ul>
<li class="chapter" data-level="2.1" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html#data-1"><i class="fa fa-check"></i><b>2.1</b> Data</a></li>
<li class="chapter" data-level="2.2" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html#create-sce"><i class="fa fa-check"></i><b>2.2</b> Create sce</a></li>
<li class="chapter" data-level="2.3" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html#summary"><i class="fa fa-check"></i><b>2.3</b> Summary</a></li>
<li class="chapter" data-level="2.4" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html#variance"><i class="fa fa-check"></i><b>2.4</b> Variance</a></li>
<li class="chapter" data-level="2.5" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html#dimension-reduction"><i class="fa fa-check"></i><b>2.5</b> Dimension reduction</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html"><i class="fa fa-check"></i><b>3</b> scM&amp;T-seq analysis</a><ul>
<li class="chapter" data-level="3.1" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html#correlation"><i class="fa fa-check"></i><b>3.1</b> Correlation</a><ul>
<li class="chapter" data-level="3.1.1" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html#pearson-correlation-study-of-methylation-rate-and-expression---gene-wise"><i class="fa fa-check"></i><b>3.1.1</b> Pearson correlation study of methylation rate and expression - gene wise</a></li>
<li class="chapter" data-level="3.1.2" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html#does-the-variation-in-methylation-in-genebody-correlates-with-variation-in-expression"><i class="fa fa-check"></i><b>3.1.2</b> Does the variation in methylation in genebody correlates with variation in expression?</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html#canonical-correlation-study"><i class="fa fa-check"></i><b>3.2</b> Canonical correlation study</a><ul>
<li class="chapter" data-level="3.2.1" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html#trascriptome---promoter-methylome"><i class="fa fa-check"></i><b>3.2.1</b> Trascriptome - Promoter Methylome</a></li>
<li class="chapter" data-level="3.2.2" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html#run-using-highly-variable-genes"><i class="fa fa-check"></i><b>3.2.2</b> Run using highly variable genes</a></li>
<li class="chapter" data-level="3.2.3" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html#more-strict-filtering-by-coverage"><i class="fa fa-check"></i><b>3.2.3</b> More strict filtering by coverage</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html#outlier-cells-diagnostics"><i class="fa fa-check"></i><b>3.3</b> Outlier cells diagnostics:</a></li>
<li class="chapter" data-level="3.4" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html#block-spls"><i class="fa fa-check"></i><b>3.4</b> Block sPLS</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/AJABADI/scMTseq-AML/" target="blank">See on GitHub</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">scMTseq of AML cells - pilot</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="scrna-seq-data" class="section level1">
<h1><span class="header-section-number">Chapter 1</span> scRNA-seq Data</h1>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## libraries</span>
<span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(mixOmics))
<span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(SingleCellExperiment))
<span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(data.table))
<span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(scran))
<span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(scater))
<span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(DESeq2))
<span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(edgeR))
<span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(Rtsne))
<span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(ggplot2)); <span class="kw">theme_set</span>(<span class="kw">theme_bw</span>())
<span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(ggrepel))
<span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(gridExtra))
<span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(grid))
<span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(reshape2))
<span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(umap))
<span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(magrittr))</code></pre>
<div id="load-raw-data" class="section level2">
<h2><span class="header-section-number">1.1</span> Load Raw Data</h2>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># system.time({</span>
        <span class="co">##  user  system elapsed </span>
        <span class="co">## 1.025   0.068   1.152 </span>
<span class="co">## read raw scRNAseq counts</span>
counts.raw =<span class="st"> </span><span class="kw">read.csv</span>(io<span class="op">$</span>rna_file)
<span class="co">## create rownames using the first column</span>
<span class="kw">rownames</span>(counts.raw) =<span class="st"> </span>counts.raw[,<span class="dv">1</span>]
<span class="co">## remove the &#39;id&#39; column</span>
counts.raw =<span class="st"> </span>counts.raw[,<span class="op">-</span><span class="dv">1</span>]
<span class="kw">head</span>(<span class="kw">colnames</span>(counts.raw))
<span class="co">## change the cell names based on well names</span>
<span class="kw">colnames</span>(counts.raw) =<span class="st"> </span><span class="kw">name.cells.by.wells</span>(<span class="kw">colnames</span>(counts.raw), <span class="dt">pos.in.string =</span> <span class="st">&#39;middle&#39;</span>)
<span class="kw">head</span>(<span class="kw">colnames</span>(counts.raw))
<span class="co">## create sce object from raw counts</span>
rna.sce =<span class="st"> </span><span class="kw">SingleCellExperiment</span>(<span class="dt">assays =</span> <span class="kw">list</span>(<span class="dt">counts =</span> <span class="kw">as.matrix</span>(counts.raw)))
rna.sce.match =<span class="st"> </span>rna.sce[,met.cellnames] <span class="co">## matching cells</span>
<span class="co"># })</span></code></pre>
</div>
<div id="preprocessing" class="section level2">
<h2><span class="header-section-number">1.2</span> Preprocessing</h2>
<div id="qc-metrics" class="section level3">
<h3><span class="header-section-number">1.2.1</span> QC metrics</h3>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## add QC metrics</span>
rna.sce =<span class="st"> </span>scater<span class="op">::</span><span class="kw">calculateQCMetrics</span>(rna.sce)</code></pre>
<div id="cell-qc" class="section level4">
<h4><span class="header-section-number">1.2.1.1</span> cell QC</h4>
<p>look at the histograms of library sizes and the number of genes expressed by each cell:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## look at the histograms of lib. sizes and the number of genes expressed by each cell</span>
<span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))
<span class="kw">hist</span>(rna.sce<span class="op">$</span>total_counts<span class="op">/</span><span class="fl">1e6</span>, <span class="dt">xlab=</span><span class="st">&quot;Library sizes (millions)&quot;</span>, <span class="dt">main=</span><span class="st">&quot;&quot;</span>,
    <span class="dt">breaks=</span><span class="dv">20</span>, <span class="dt">col=</span><span class="st">&quot;grey80&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;Number of cells&quot;</span>)
<span class="kw">hist</span>(rna.sce<span class="op">$</span>total_features_by_counts, <span class="dt">xlab=</span><span class="st">&quot;Number of expressed genes&quot;</span>, <span class="dt">main=</span><span class="st">&quot;&quot;</span>,
    <span class="dt">breaks=</span><span class="dv">20</span>, <span class="dt">col=</span><span class="st">&quot;grey80&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;Number of cells&quot;</span>)</code></pre>
<p><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-8-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Create a list of cell QC metrics:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## keep a record of lower limit cells</span>
cell.qc =<span class="st"> </span><span class="kw">list</span>()

<span class="co">## qc metric based on lib size</span>
tot.counts =<span class="st"> </span>rna.sce<span class="op">$</span>total_counts
<span class="kw">names</span>(tot.counts) =<span class="st"> </span><span class="kw">colnames</span>(rna.sce)
tot.counts =<span class="st"> </span>tot.counts[<span class="kw">order</span>(tot.counts)] <span class="co">## from least to most</span>
libsize =<span class="st"> </span><span class="kw">list</span>()
libsize<span class="op">$</span>rank =<span class="st"> </span><span class="kw">rank</span>(tot.counts)
libsize<span class="op">$</span>ratio2median =<span class="st"> </span>tot.counts<span class="op">/</span><span class="kw">median</span>(tot.counts)
cell.qc<span class="op">$</span>libsize =<span class="st"> </span>libsize

<span class="co">## qc metric based on gene expression</span>
genes.expressed.ratio2median =<span class="st"> </span>rna.sce<span class="op">$</span>total_features_by_counts<span class="op">/</span><span class="kw">median</span>(rna.sce<span class="op">$</span>total_features_by_counts)
<span class="kw">names</span>(genes.expressed.ratio2median) =<span class="st"> </span><span class="kw">colnames</span>(rna.sce)
genes.expressed.ratio2median <span class="op">%&gt;%</span><span class="st"> </span>.[<span class="kw">order</span>(.)] <span class="co">## from least to most</span>
cell.qc<span class="op">$</span>gene.expressed.ratio2median =<span class="st"> </span>genes.expressed.ratio2median</code></pre>
<p>Which cells are more than 3.5 median absolute deviations less than the median log library size / number of genes expressed?:</p>
<pre class="sourceCode r"><code class="sourceCode r">libsize.drop &lt;-<span class="st"> </span><span class="kw">isOutlier</span>(rna.sce<span class="op">$</span>total_counts, <span class="dt">nmads=</span><span class="fl">3.5</span>, <span class="dt">type=</span><span class="st">&quot;lower&quot;</span>, <span class="dt">log=</span><span class="ot">TRUE</span>)
<span class="kw">colnames</span>(rna.sce)[libsize.drop] <span class="co">## the cells with low library size</span>
feature.drop &lt;-<span class="st"> </span><span class="kw">isOutlier</span>(rna.sce<span class="op">$</span>total_features_by_counts, <span class="dt">nmads=</span><span class="fl">3.5</span>, <span class="dt">type=</span><span class="st">&quot;lower&quot;</span>, <span class="dt">log=</span><span class="ot">TRUE</span>)
<span class="kw">colnames</span>(rna.sce)[feature.drop] <span class="co">## the cells with low number of expressed genes &gt; NULL</span></code></pre>
<p>For each cell, we look at number of genes expressed vs library size (ratio to the median for both):</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## cells</span>
df4ggplot =<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">libsize =</span> libsize<span class="op">$</span>ratio2median,
                <span class="dt">genes.expressed =</span> genes.expressed.ratio2median[<span class="kw">names</span>(libsize<span class="op">$</span>ratio2median)])
<span class="co">## keep a record of lower quality cells for future reference</span>

rank.threshold =<span class="st"> </span><span class="dv">5</span> <span class="co">## the rank threshold on library size and gene expression</span>

df4ggplot<span class="op">$</span>lower.quality =<span class="st"> </span><span class="kw">with</span>(df4ggplot, <span class="kw">rank</span>(libsize)<span class="op">&lt;</span>rank.threshold <span class="op">|</span>
<span class="st">                           </span><span class="kw">rank</span>(genes.expressed)<span class="op">&lt;</span>rank.threshold)

<span class="kw">ggplot</span>(df4ggplot, <span class="kw">aes</span>(<span class="dt">x =</span> libsize, <span class="dt">y =</span> genes.expressed)) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">size =</span> <span class="dv">3</span>, <span class="dt">col=</span><span class="kw">color.mixo</span>(<span class="dv">1</span>)) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&#39;Library Size / median&#39;</span> , <span class="dt">y=</span><span class="st">&#39; Number of genes expressed /median&#39;</span>, <span class="dt">title =</span> <span class="st">&#39;# of genes expressed against library size for all cells. </span><span class="ch">\n</span><span class="st"> cells on lower limits are labelled for downstream reference&#39;</span>) <span class="op">+</span>
<span class="st">   </span><span class="kw">theme</span>(<span class="dt">plot.title =</span> <span class="kw">element_text</span>(<span class="dt">hjust =</span> <span class="fl">0.5</span>)) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_label_repel</span>(
    <span class="kw">aes</span>(<span class="dt">label =</span> <span class="kw">ifelse</span>(lower.quality,<span class="kw">rownames</span>(df4ggplot),<span class="st">&#39;&#39;</span>)),
    <span class="dt">col =</span> <span class="st">&#39;darkorange&#39;</span>,
    <span class="dt">fontface =</span> <span class="st">&#39;bold&#39;</span>,
    <span class="dt">box.padding =</span> <span class="kw">unit</span>(<span class="fl">0.35</span>, <span class="st">&quot;lines&quot;</span>),
    <span class="dt">point.padding =</span> <span class="kw">unit</span>(<span class="fl">0.5</span>, <span class="st">&quot;lines&quot;</span>),
    <span class="dt">segment.color =</span> <span class="st">&#39;grey50&#39;</span>
  )</code></pre>
<p><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-11-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>No cells are filtered at this stage. We will assess <strong>A3</strong> and <strong>E8</strong> further downstream.</p>
</div>
<div id="gene-filtering" class="section level4">
<h4><span class="header-section-number">1.2.1.2</span> gene filtering</h4>
<p>We require genes to have:
+ non-zero library size
+ be expressed/detected in at least 5% of cells</p>
<pre class="sourceCode r"><code class="sourceCode r">gene.metrics =<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">rowData</span>(rna.sce)) <span class="co">## a data.frame of gene metrics</span>
gene.metrics<span class="op">$</span>pass &lt;-<span class="st"> </span>gene.metrics<span class="op">$</span>total_counts<span class="op">!=</span><span class="dv">0</span> <span class="op">&amp;</span><span class="st"> </span>gene.metrics<span class="op">$</span>pct_dropout_by_counts<span class="op">&lt;</span>params<span class="op">$</span>max.cell.dropout<span class="op">*</span><span class="dv">100</span>
<span class="kw">sum</span>(gene.metrics<span class="op">$</span>pass) <span class="co">## number of genes after this filter: 15K</span></code></pre>
<p>To also filter out noisy genes with high rate of dropouts, we usually look at the library size histogram and put a threshold between the two peaks to the left of where the fitted curve plateaues <span class="citation">(Lun, McCarthy, and Marioni 2018)</span>.</p>
<!-- This can be mathematically calculated by fitting a mixture model (gamma normal maybe?) to the curve and calculating the minimum between the two maxima. -->
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## choose a min log lib size threshold</span>
min.libsize =<span class="st"> </span><span class="kw">log10</span>(<span class="dv">4</span>) <span class="co">## must be chosen based on the following hisogram</span>
<span class="co">## look at only those which passed total lib size filtering and cell dropout filtering</span>
log.mean.counts =<span class="st"> </span>gene.metrics<span class="op">$</span>log10_mean_counts
<span class="co">## see if you&#39;re filtering the first peak of non-zero values</span>
hist =<span class="st"> </span><span class="kw">hist</span>(log.mean.counts[log.mean.counts<span class="op">&gt;</span><span class="dv">0</span>], <span class="dt">prob=</span><span class="ot">TRUE</span>, <span class="dt">breaks =</span> <span class="dv">100</span>, <span class="dt">col =</span> <span class="st">&#39;grey80&#39;</span>,
            <span class="dt">main=</span> <span class="st">&#39;Histogram of log10 mean gene counts / library sizes&#39;</span>, <span class="dt">xlab =</span> <span class="st">&#39;log10[mean counts] for each gene&#39;</span>)
<span class="kw">abline</span>(<span class="dt">v =</span> min.libsize, <span class="dt">col =</span> <span class="st">&quot;darkorange&quot;</span>, <span class="dt">lwd=</span><span class="dv">2</span>)</code></pre>
<p><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-13-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Based on the histogam above, 0.60206 is chosen as minimum gene library size.</p>
<!-- ```{r, eval=FALSE, echo=FALSE, purl = F} -->
<!-- ## fit a Guassian mixture model (k=2) -->
<!-- fit = mixtools::normalmixEM(log10libsize, mu = c(0.2, 1.4), k=2) -->
<!-- ## define the density function -->
<!-- dens.fun = function(x, fit=fit){ -->
<!-- fit$lambda[1]*dnorm(x, fit$mu[1], fit$sigma[1]) + -->
<!-- fit$lambda[2]*dnorm(x, fit$mu[2], fit$sigma[2]) -->
<!-- } -->
<!-- ## plot -->
<!-- curve( expr = fit$lambda[1]*dnorm(x, fit$mu[1], fit$sigma[1]) + -->
<!--          fit$lambda[2]*dnorm(x, fit$mu[2], fit$sigma[2]), 0, (fit$mu[2] + 3*fit$sigma[2])) -->
<!-- ## get the maximum and the local minimum -->
<!-- ymax = optimize(f = function(x) (dens(x,fit=fit)),interval = c(-10,10), maximum = T) -->
<!-- ymin = optimize(f = function(x) (dens(x,fit=fit)),interval = c(ymax$maximum,0.7), maximum = F) -->
<!-- ``` -->
<pre class="sourceCode r"><code class="sourceCode r">gene.metrics<span class="op">$</span>pass[gene.metrics<span class="op">$</span>pass] &lt;-<span class="st"> </span>gene.metrics<span class="op">$</span>log10_mean_counts[
  gene.metrics<span class="op">$</span>pass]<span class="op">&gt;=</span>params<span class="op">$</span>min.gene.libsize
<span class="kw">sum</span>(gene.metrics<span class="op">$</span>pass) <span class="co">## number of genes after this filter: 10.7K</span></code></pre>
<pre class="sourceCode r"><code class="sourceCode r">rna_qc =<span class="st"> </span>rna.sce[gene.metrics<span class="op">$</span>pass,] <span class="co">## filter genes</span></code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(rna_qc) <span class="co">## 10778    78</span></code></pre>
</div>
</div>
</div>
<div id="normalisation" class="section level2">
<h2><span class="header-section-number">1.3</span> Normalisation</h2>
<div id="scran" class="section level3">
<h3><span class="header-section-number">1.3.1</span> scran</h3>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## use scran to pool cells based on gene expression similarity and calculate size factors</span>
<span class="kw">system.time</span>({
  <span class="co">##    user  system elapsed </span>
   <span class="co">## 0.366   0.027   0.435 </span>
rna_qc =<span class="st"> </span><span class="kw">computeSumFactors</span>(rna_qc)
<span class="co">## normalise using calculated size factors</span>
rna_qc =<span class="st"> </span><span class="kw">normalize</span>(rna_qc)
})</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## size factors summary</span>
<span class="co"># summary(sizeFactors(rna_qc))</span>
<span class="kw">plot</span>(<span class="kw">sizeFactors</span>(rna_qc), rna_qc<span class="op">$</span>total_counts<span class="op">/</span><span class="fl">1e3</span>, <span class="dt">log=</span><span class="st">&quot;xy&quot;</span>, <span class="dt">pch=</span><span class="dv">16</span>, <span class="dt">col=</span><span class="st">&#39;purple&#39;</span>,
    <span class="dt">ylab=</span><span class="st">&quot;Library size (thousands)&quot;</span>, <span class="dt">xlab=</span><span class="st">&quot;Size factor&quot;</span>, <span class="dt">main =</span> <span class="st">&quot;Scran Size Factors&quot;</span>)</code></pre>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-17"></span>
<img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-17-1.png" alt="Size factors from deconvolution, plotted against library sizes for all cells. Axes are shown on a log-scale. Since we have similar cell types (and not much DE genes) we expect to see less scatter around the trend line, whcih is the case in this plot." width="672" />
<p class="caption">
Figure 1.1: Size factors from deconvolution, plotted against library sizes for all cells. Axes are shown on a log-scale. Since we have similar cell types (and not much DE genes) we expect to see less scatter around the trend line, whcih is the case in this plot.
</p>
</div>
</div>
</div>
<div id="dimension-reduction---filtered-genes" class="section level2">
<h2><span class="header-section-number">1.4</span> Dimension reduction - filtered genes</h2>
<div id="pca" class="section level3">
<h3><span class="header-section-number">1.4.1</span> PCA</h3>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## scran</span>
pca.res =<span class="st"> </span><span class="kw">pca</span>(<span class="kw">t</span>(<span class="kw">logcounts</span>(rna_qc)), <span class="dt">ncomp =</span> <span class="dv">5</span>)</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">rnaseq.only =<span class="st"> </span><span class="kw">colnames</span>(rna_qc)[<span class="op">!</span><span class="st"> </span><span class="kw">colnames</span>(rna_qc) <span class="op">%in%</span><span class="st"> </span>met.cellnames]
<span class="kw">pca.grid</span>(<span class="dt">pca.obj =</span> pca.res, <span class="dt">diff.samples =</span> rnaseq.only)</code></pre>
<p><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-19-1.png" width="100%" style="display: block; margin: auto;" /></p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## invesigating whether there are technical effects by wells:</span>
<span class="kw">source</span>(<span class="st">&quot;utils/pca.well.R&quot;</span>)
<span class="kw">pca.well</span>(pca.res, <span class="dt">PCs=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</code></pre>
<p><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-20-1.png" width="100%" style="display: block; margin: auto;" /></p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">pca.well</span>(pca.res, <span class="dt">PCs=</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">3</span>))</code></pre>
<p><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-20-2.png" width="100%" style="display: block; margin: auto;" /></p>
<p>There does not seems to be a grouping by well rows from PCA plots.</p>
</div>
<div id="variation" class="section level3">
<h3><span class="header-section-number">1.4.2</span> Variation</h3>
<p>Since there are technical factors that can affect capture efficiency (such as GC content for Illumina sequencing, PCR amplification noise etc), for the studies that use HVGs, we need to ensure that are gene selection is based on biological variance.</p>
<div id="variance-decomposition-method" class="section level4">
<h4><span class="header-section-number">1.4.2.1</span> variance decomposition method</h4>
<p>By fitting a mean-depndent <em>loess</em> curve to variance - <strong>log</strong> expression (which makes it more robust to outliers).</p>
<p>The fitted value of this trend represents technical variability due to sequencing, drop-outs during capture, etc. at a given mean.</p>
</div>
<div id="how-is-biological-variance-determined" class="section level4">
<h4><span class="header-section-number">1.4.2.2</span> How is biological variance determined?</h4>
<p>Details can be found <a href="https://github.com/MarioniLab/scran/blob/master/R/trendVar.R">here</a> and <a href="https://github.com/LTLA/HVGDetection2018">here</a>.</p>
</div>
<div id="decompose-variance" class="section level4">
<h4><span class="header-section-number">1.4.2.3</span> decompose variance</h4>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## function to decompose variance for sce and return results as rowData(sce)</span>
var.decomp =<span class="st"> </span><span class="cf">function</span>(<span class="dt">sce=</span>rna_qc){
  fit &lt;-<span class="st"> </span><span class="kw">trendVar</span>(sce, <span class="dt">method=</span><span class="st">&quot;loess&quot;</span>, <span class="dt">use.spikes=</span><span class="ot">FALSE</span>)  <span class="co">## fit a mean-dependent loess to variance</span>
  decomp &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">decomposeVar</span>(sce, fit)) <span class="co">## decompose variance to technical and biological</span>
  decomp<span class="op">$</span>trend.var &lt;-<span class="st"> </span>fit<span class="op">$</span><span class="kw">trend</span>(fit<span class="op">$</span>means) <span class="co">## add the fitted variance column</span>
  <span class="kw">rowData</span>(sce) <span class="op">%&lt;&gt;%</span><span class="st">  </span><span class="kw">cbind</span>(.,decomp) <span class="co">## add to rowData(sce)</span>
  sce <span class="op">%&lt;&gt;%</span><span class="st"> </span>.[<span class="kw">order</span>(<span class="op">-</span><span class="kw">rowData</span>(.)<span class="op">$</span>bio),] <span class="co">## order by bio variance</span>
  
  <span class="kw">return</span>(sce)
}</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## decompose variance</span>
rna_qc <span class="op">%&lt;&gt;%</span><span class="st"> </span><span class="kw">var.decomp</span>(.)</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))
row_data =<span class="st">  </span>rna_qc <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">rowData</span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as.data.frame</span>() <span class="co">## a data.frame of rowData</span>
<span class="co">## total variance (fold change compared to mean) vs log mean</span>
<span class="kw">ggplot</span>(row_data) <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(<span class="dt">x=</span> mean, <span class="dt">y=</span>total, <span class="dt">col =</span> FDR), <span class="dt">show.legend =</span> <span class="ot">TRUE</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">scale_colour_gradientn</span>(<span class="dt">colors =</span> <span class="kw">c</span>(<span class="st">&quot;#027009&quot;</span>, <span class="st">&quot;#aebc31&quot;</span>), <span class="dt">name =</span> <span class="st">&quot;FDR&quot;</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_line</span>(<span class="kw">aes</span>(<span class="dt">x=</span>mean, <span class="dt">y =</span> trend.var, <span class="dt">fill =</span> <span class="st">&#39;Fitted trend&#39;</span>),  <span class="dt">size=</span><span class="dv">2</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">scale_fill_manual</span>(<span class="st">&quot;&quot;</span>, <span class="dt">values =</span> <span class="kw">c</span>(<span class="dv">6</span>), <span class="dt">guide=</span><span class="kw">guide_legend</span>(<span class="dt">override.aes =</span> <span class="kw">list</span>(<span class="dt">color=</span><span class="kw">c</span>(<span class="st">&quot;black&quot;</span>)))) <span class="op">+</span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">x=</span><span class="st">&#39;mean log expression&#39;</span>, <span class="dt">y =</span> <span class="st">&#39;total variance&#39;</span>, <span class="dt">title =</span> <span class="st">&#39; total variance vs log mean expression for filtered genes&#39;</span>)</code></pre>
<p><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-23-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## inferred biological variance (fold change compared to mean) vs log mean</span>
<span class="kw">ggplot</span>(row_data) <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(<span class="dt">x=</span> mean, <span class="dt">y=</span>bio, <span class="dt">col=</span>FDR), <span class="dt">show.legend =</span> <span class="ot">TRUE</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">scale_colour_gradientn</span>(<span class="dt">colors =</span> <span class="kw">c</span>(<span class="st">&quot;#027009&quot;</span>, <span class="st">&quot;#aebc31&quot;</span>), <span class="dt">name =</span> <span class="st">&quot;FDR&quot;</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">x=</span><span class="st">&#39;mean log expression&#39;</span>, <span class="dt">y =</span> <span class="st">&#39;biological variance&#39;</span>,
       <span class="dt">title =</span> <span class="st">&#39; biological variation vs log mean expression for filtered genes&#39;</span>)</code></pre>
<p><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-23-2.png" width="672" style="display: block; margin: auto;" /></p>
<div id="pearsons-r" class="section level5">
<h5><span class="header-section-number">1.4.2.3.1</span> Pearson’s <em>r</em></h5>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## hierarchical clustering using HVGs</span>
hvgs =<span class="st"> </span><span class="kw">rownames</span>(rna_qc[<span class="dv">1</span><span class="op">:</span>params<span class="op">$</span>rna<span class="fl">.78</span>cells.hvgs,])
<span class="kw">plotHeatmap</span>(rna_qc[hvgs,], <span class="dt">features =</span> hvgs, <span class="dt">clutering_distance_rows =</span> <span class="st">&#39;correlation&#39;</span>, <span class="dt">clutering_distance_cols =</span> <span class="st">&#39;correlation&#39;</span>, <span class="dt">main =</span> <span class="kw">paste0</span>( <span class="st">&#39;Clustered heatmap of scRNAseq data with &#39;</span>,params<span class="op">$</span>rna<span class="fl">.78</span>cells.hvgs, <span class="st">&#39; HVGs only </span><span class="ch">\n</span><span class="st"> using Pearson correlation </span><span class="ch">\n</span><span class="st"> cells at the bottom and genes on the side&#39;</span>))</code></pre>
<p><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-24-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
</div>
</div>
</div>
<div id="dimension-reduction---hvgs" class="section level2">
<h2><span class="header-section-number">1.5</span> Dimension reduction - HVGs</h2>
<div id="pca-using-hvgs" class="section level3">
<h3><span class="header-section-number">1.5.1</span> PCA using HVGS</h3>
<p>The first 2 PCs now contain ~ 13% of variation, compared to 5% which was before</p>
<pre class="sourceCode r"><code class="sourceCode r">pca.hvgs.res =<span class="st"> </span><span class="kw">pca</span>(<span class="kw">t</span>(<span class="kw">logcounts</span>(rna_qc[hvgs,])), <span class="dt">ncomp =</span> <span class="dv">5</span>)</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">pca.grid</span>(<span class="dt">pca.obj =</span> pca.hvgs.res, <span class="dt">diff.samples =</span> rnaseq.only)</code></pre>
<p><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-26-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<div id="tsne-does-not-identify-clear-clusters" class="section level3">
<h3><span class="header-section-number">1.5.2</span> tSNE does not identify clear clusters</h3>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))
<span class="kw">set.seed</span>(<span class="dv">12321</span>)
sce.hvg.tsne =<span class="st"> </span><span class="kw">Rtsne</span>(<span class="kw">t</span>(<span class="kw">logcounts</span>(rna_qc[hvgs,])), <span class="dt">perplexity =</span> <span class="dv">10</span>)</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(sce.hvg.tsne<span class="op">$</span>Y, <span class="dt">main =</span> <span class="st">&#39;tSNE - Perplexity = 10&#39;</span>, <span class="dt">pch=</span><span class="kw">colnames</span>(rna_qc),
     <span class="dt">col=</span><span class="kw">factor</span>(<span class="kw">substr</span>(<span class="kw">colnames</span>(rna_qc),<span class="dv">0</span>,<span class="dv">1</span>)), <span class="dt">xlab =</span> <span class="st">&#39;tSNE-1&#39;</span>, <span class="dt">ylab=</span><span class="st">&#39;tSNE-2&#39;</span>)</code></pre>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-28"></span>
<img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-28-1.png" alt="tSNE plots using HVGs" width="672" />
<p class="caption">
Figure 1.2: tSNE plots using HVGs
</p>
</div>
</div>
<div id="umap-does-not-identify-clear-clusters" class="section level3">
<h3><span class="header-section-number">1.5.3</span> UMAP does not identify clear clusters</h3>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co">## UMAP on scran-normalised</span>
sce.hvg.umap =<span class="st"> </span><span class="kw">umap</span>(<span class="kw">t</span>(<span class="kw">logcounts</span>(rna_qc[hvgs,])))</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(sce.hvg.umap<span class="op">$</span>layout, <span class="dt">main =</span> <span class="st">&#39;UMAP using HVGs&#39;</span>, <span class="dt">pch=</span><span class="kw">rownames</span>(sce.hvg.umap<span class="op">$</span>layout),
     <span class="dt">col=</span><span class="kw">factor</span>(<span class="kw">substr</span>(<span class="kw">rownames</span>(sce.hvg.umap<span class="op">$</span>layout),<span class="dv">0</span>,<span class="dv">1</span>)), <span class="dt">xlab =</span> <span class="st">&#39;umap-1&#39;</span>, <span class="dt">ylab=</span><span class="st">&#39;umap-2&#39;</span> )</code></pre>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-30"></span>
<img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-30-1.png" alt="UMAP plot using HVGs colored by wells" width="672" />
<p class="caption">
Figure 1.3: UMAP plot using HVGs colored by wells
</p>
</div>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="scbs-seq-data.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
