<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>scMTseq of AML cells - pilot</title>
  <meta name="description" content="Joint methylation and transcriptome sequencing of AML cells">
  <meta name="generator" content="bookdown 0.7 and GitBook 2.6.7">

  <meta property="og:title" content="scMTseq of AML cells - pilot" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Joint methylation and transcriptome sequencing of AML cells" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="scMTseq of AML cells - pilot" />
  
  <meta name="twitter:description" content="Joint methylation and transcriptome sequencing of AML cells" />
  

<meta name="author" content="Al J Abadi">


<meta name="date" content="2018-11-29">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="index.html">
<link rel="next" href="scbs-seq-data.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Pilot scMTseq of AML cells</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Data</a></li>
<li class="chapter" data-level="1" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html"><i class="fa fa-check"></i><b>1</b> scRNA-seq Data</a><ul>
<li class="chapter" data-level="1.1" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#load-raw-data"><i class="fa fa-check"></i><b>1.1</b> Load Raw Data</a></li>
<li class="chapter" data-level="1.2" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#preprocessing"><i class="fa fa-check"></i><b>1.2</b> Preprocessing</a><ul>
<li class="chapter" data-level="1.2.1" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#qc-metrics"><i class="fa fa-check"></i><b>1.2.1</b> QC metrics</a></li>
<li class="chapter" data-level="1.2.2" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#cell-qc"><i class="fa fa-check"></i><b>1.2.2</b> Cell QC</a></li>
<li class="chapter" data-level="1.2.3" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#detect-outliers"><i class="fa fa-check"></i><b>1.2.3</b> Detect outliers</a></li>
<li class="chapter" data-level="1.2.4" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#gene-filtering"><i class="fa fa-check"></i><b>1.2.4</b> Gene Filtering</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#normalization"><i class="fa fa-check"></i><b>1.3</b> Normalization</a><ul>
<li class="chapter" data-level="1.3.1" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#scran"><i class="fa fa-check"></i><b>1.3.1</b> scran</a></li>
<li class="chapter" data-level="1.3.2" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#tmm-edger"><i class="fa fa-check"></i><b>1.3.2</b> TMM (edgeR)</a></li>
<li class="chapter" data-level="1.3.3" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#deseq2"><i class="fa fa-check"></i><b>1.3.3</b> DESeq2</a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#dimension-reduction"><i class="fa fa-check"></i><b>1.4</b> Dimension reduction</a><ul>
<li class="chapter" data-level="1.4.1" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#pca"><i class="fa fa-check"></i><b>1.4.1</b> PCA</a></li>
<li class="chapter" data-level="1.4.2" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#hierarchical-clustering"><i class="fa fa-check"></i><b>1.4.2</b> Hierarchical Clustering</a></li>
<li class="chapter" data-level="1.4.3" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#tsne"><i class="fa fa-check"></i><b>1.4.3</b> tSNE</a></li>
<li class="chapter" data-level="1.4.4" data-path="scrna-seq-data.html"><a href="scrna-seq-data.html#looking-at-umap"><i class="fa fa-check"></i><b>1.4.4</b> Looking at UMAP</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html"><i class="fa fa-check"></i><b>2</b> scBS-seq Data</a><ul>
<li class="chapter" data-level="2.1" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html#methylation-and-rnaseq"><i class="fa fa-check"></i><b>2.1</b> Methylation and RNAseq</a><ul>
<li class="chapter" data-level="2.1.1" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html#filtering-nas"><i class="fa fa-check"></i><b>2.1.1</b> Filtering NAs</a></li>
<li class="chapter" data-level="2.1.2" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html#hvgs-in-methylation-data"><i class="fa fa-check"></i><b>2.1.2</b> HVGs in Methylation Data</a></li>
<li class="chapter" data-level="2.1.3" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html#spls---gene-body"><i class="fa fa-check"></i><b>2.1.3</b> sPLS - Gene Body</a></li>
<li class="chapter" data-level="2.1.4" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html#spls---promoter"><i class="fa fa-check"></i><b>2.1.4</b> sPLS - Promoter</a></li>
<li class="chapter" data-level="2.1.5" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html#block-spls"><i class="fa fa-check"></i><b>2.1.5</b> Block sPLS</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html#kb-windows"><i class="fa fa-check"></i><b>2.2</b> 3kb Windows</a><ul>
<li class="chapter" data-level="2.2.1" data-path="scbs-seq-data.html"><a href="scbs-seq-data.html#dimension-reduction-1"><i class="fa fa-check"></i><b>2.2.1</b> Dimension Reduction</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="scmt-seq-analysis.html"><a href="scmt-seq-analysis.html"><i class="fa fa-check"></i><b>3</b> scM&amp;T-seq analysis</a></li>
<li class="divider"></li>
<li><a href="https://github.com/AJABADI/scMTseq-AML/" target="blank">See on GitHub</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">scMTseq of AML cells - pilot</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="scrna-seq-data" class="section level1">
<h1><span class="header-section-number">Chapter 1</span> scRNA-seq Data</h1>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="co">## libraries</span></a>
<a class="sourceLine" id="cb5-2" data-line-number="2"><span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(mixOmics))</a>
<a class="sourceLine" id="cb5-3" data-line-number="3"><span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(SingleCellExperiment))</a>
<a class="sourceLine" id="cb5-4" data-line-number="4"><span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(data.table))</a>
<a class="sourceLine" id="cb5-5" data-line-number="5"><span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(scran))</a>
<a class="sourceLine" id="cb5-6" data-line-number="6"><span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(scater))</a>
<a class="sourceLine" id="cb5-7" data-line-number="7"><span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(DESeq2))</a>
<a class="sourceLine" id="cb5-8" data-line-number="8"><span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(edgeR))</a>
<a class="sourceLine" id="cb5-9" data-line-number="9"><span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(Rtsne))</a>
<a class="sourceLine" id="cb5-10" data-line-number="10"><span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(ggplot2)); <span class="kw">theme_set</span>(<span class="kw">theme_bw</span>())</a>
<a class="sourceLine" id="cb5-11" data-line-number="11"><span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(ggrepel))</a>
<a class="sourceLine" id="cb5-12" data-line-number="12"><span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(gridExtra))</a>
<a class="sourceLine" id="cb5-13" data-line-number="13"><span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(grid))</a>
<a class="sourceLine" id="cb5-14" data-line-number="14"><span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(reshape2))</a></code></pre></div>
<div id="load-raw-data" class="section level2">
<h2><span class="header-section-number">1.1</span> Load Raw Data</h2>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="co"># system.time({</span></a>
<a class="sourceLine" id="cb6-2" data-line-number="2">        <span class="co">##  user  system elapsed </span></a>
<a class="sourceLine" id="cb6-3" data-line-number="3">        <span class="co">## 1.025   0.068   1.152 </span></a>
<a class="sourceLine" id="cb6-4" data-line-number="4"><span class="co">## read raw scRNAseq counts</span></a>
<a class="sourceLine" id="cb6-5" data-line-number="5">counts.raw =<span class="st"> </span><span class="kw">read.csv</span>(io<span class="op">$</span>rna_file)</a>
<a class="sourceLine" id="cb6-6" data-line-number="6"><span class="co">## create rownames using the first column</span></a>
<a class="sourceLine" id="cb6-7" data-line-number="7"><span class="kw">rownames</span>(counts.raw) =<span class="st"> </span>counts.raw[,<span class="dv">1</span>]</a>
<a class="sourceLine" id="cb6-8" data-line-number="8">counts.raw =<span class="st"> </span>counts.raw[,<span class="op">-</span><span class="dv">1</span>]</a>
<a class="sourceLine" id="cb6-9" data-line-number="9"><span class="kw">head</span>(<span class="kw">colnames</span>(counts.raw))</a></code></pre></div>
<pre><code>## [1] &quot;X180924_A10_S10.bam&quot; &quot;X180924_A2_S2.bam&quot;   &quot;X180924_A3_S3.bam&quot;  
## [4] &quot;X180924_A4_S4.bam&quot;   &quot;X180924_A5_S5.bam&quot;   &quot;X180924_A6_S6.bam&quot;</code></pre>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="co">## change the cell names based on well names</span></a>
<a class="sourceLine" id="cb8-2" data-line-number="2"><span class="kw">source</span>(<span class="st">&#39;utils//name.cells.by.wells.R&#39;</span>)</a>
<a class="sourceLine" id="cb8-3" data-line-number="3"><span class="kw">colnames</span>(counts.raw) =<span class="st"> </span><span class="kw">name.cells.by.wells</span>(<span class="kw">colnames</span>(counts.raw), <span class="dt">pos.in.string =</span> <span class="st">&#39;middle&#39;</span>)</a>
<a class="sourceLine" id="cb8-4" data-line-number="4"><span class="kw">head</span>(<span class="kw">colnames</span>(counts.raw))</a></code></pre></div>
<pre><code>## [1] &quot;A10&quot; &quot;A2&quot;  &quot;A3&quot;  &quot;A4&quot;  &quot;A5&quot;  &quot;A6&quot;</code></pre>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"><span class="co">## create sce object from raw counts</span></a>
<a class="sourceLine" id="cb10-2" data-line-number="2">sce.rna =<span class="st"> </span><span class="kw">SingleCellExperiment</span>(<span class="dt">assays =</span> <span class="kw">list</span>(<span class="dt">counts =</span> <span class="kw">as.matrix</span>(counts.raw)))</a>
<a class="sourceLine" id="cb10-3" data-line-number="3"><span class="co"># })</span></a></code></pre></div>
</div>
<div id="preprocessing" class="section level2">
<h2><span class="header-section-number">1.2</span> Preprocessing</h2>
<div id="qc-metrics" class="section level3">
<h3><span class="header-section-number">1.2.1</span> QC metrics</h3>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="co">## add QC metrics</span></a>
<a class="sourceLine" id="cb11-2" data-line-number="2">sce.rna =<span class="st"> </span>scater<span class="op">::</span><span class="kw">calculateQCMetrics</span>(sce.rna)</a></code></pre></div>
</div>
<div id="cell-qc" class="section level3">
<h3><span class="header-section-number">1.2.2</span> Cell QC</h3>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1"><span class="co">## show lib. size and the gene expression histograms</span></a>
<a class="sourceLine" id="cb12-2" data-line-number="2"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</a>
<a class="sourceLine" id="cb12-3" data-line-number="3"><span class="kw">hist</span>(sce.rna<span class="op">$</span>total_counts<span class="op">/</span><span class="fl">1e3</span>, <span class="dt">xlab=</span><span class="st">&quot;Library sizes (thousands)&quot;</span>, <span class="dt">main=</span><span class="st">&quot;&quot;</span>, </a>
<a class="sourceLine" id="cb12-4" data-line-number="4">     <span class="dt">breaks=</span><span class="dv">20</span>, <span class="dt">col=</span><span class="st">&quot;grey80&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;Number of cells&quot;</span>)</a>
<a class="sourceLine" id="cb12-5" data-line-number="5"><span class="kw">hist</span>(sce.rna<span class="op">$</span>total_features_by_counts, <span class="dt">xlab=</span><span class="st">&quot;Number of expressed genes&quot;</span>, <span class="dt">main=</span><span class="st">&quot;&quot;</span>, </a>
<a class="sourceLine" id="cb12-6" data-line-number="6">     <span class="dt">breaks=</span><span class="dv">20</span>, <span class="dt">col=</span><span class="st">&quot;grey80&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;Number of cells&quot;</span>)</a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:libsize"></span>
<img src="scMTseq_of_AML_cells_files/figure-html/libsize-1.png" alt="Histograms of cell QC metrics including the library sizes and number of expressed genes" width="100%" />
<p class="caption">
Figure 1.1: Histograms of cell QC metrics including the library sizes and number of expressed genes
</p>
</div>
</div>
<div id="detect-outliers" class="section level3">
<h3><span class="header-section-number">1.2.3</span> Detect outliers</h3>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1"><span class="co">## lib size too small - 3.5 median abs. deviation less than median</span></a>
<a class="sourceLine" id="cb13-2" data-line-number="2"><span class="co">## chose 3.5 MAD instead of 3 as we don&#39;t want to lose the precious cells!</span></a>
<a class="sourceLine" id="cb13-3" data-line-number="3">libsize.drop =<span class="st"> </span><span class="kw">isOutlier</span>(sce.rna<span class="op">$</span>total_counts, <span class="dt">nmads=</span><span class="fl">3.5</span>, <span class="dt">type=</span><span class="st">&quot;lower&quot;</span>, <span class="dt">log=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb13-4" data-line-number="4"><span class="kw">sum</span>(libsize.drop) <span class="co">## nothing dropped</span></a></code></pre></div>
<pre><code>## [1] 0</code></pre>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1"><span class="co">## expressing low genes</span></a>
<a class="sourceLine" id="cb15-2" data-line-number="2">feature.drop =<span class="st"> </span><span class="kw">isOutlier</span>(sce.rna<span class="op">$</span>total_features_by_counts, <span class="dt">nmads=</span><span class="dv">3</span>, <span class="dt">type=</span><span class="st">&quot;lower&quot;</span>, <span class="dt">log=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb15-3" data-line-number="3"><span class="kw">sum</span>(feature.drop) <span class="co">## nothing dropped</span></a></code></pre></div>
<pre><code>## [1] 0</code></pre>
<p>All cells passed QC metrics for lib size and gene expression</p>
</div>
<div id="gene-filtering" class="section level3">
<h3><span class="header-section-number">1.2.4</span> Gene Filtering</h3>
<div id="using-mean-expression" class="section level4">
<h4><span class="header-section-number">1.2.4.1</span> using mean expression</h4>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1"><span class="co">## histogram of log mean counts for all genes</span></a>
<a class="sourceLine" id="cb17-2" data-line-number="2">avg.count.cutoff =<span class="st"> </span><span class="dv">1</span> <span class="co">## for log2count = 1</span></a>
<a class="sourceLine" id="cb17-3" data-line-number="3"><span class="co">## get size factors</span></a>
<a class="sourceLine" id="cb17-4" data-line-number="4">sce.rna =<span class="st"> </span><span class="kw">computeSumFactors</span>(sce.rna)</a></code></pre></div>
<pre><code>## Warning in FUN(...): not enough cells in at least one cluster for some
## &#39;sizes&#39;</code></pre>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1">ave.counts =<span class="st"> </span><span class="kw">calcAverage</span>(sce.rna, <span class="dt">use_size_factors=</span>T)</a>
<a class="sourceLine" id="cb19-2" data-line-number="2">log2.raw =<span class="st"> </span><span class="kw">log2</span>(ave.counts)</a>
<a class="sourceLine" id="cb19-3" data-line-number="3">hist =<span class="st"> </span><span class="kw">hist</span>(log2.raw, <span class="dt">breaks=</span><span class="dv">100</span>, <span class="dt">plot =</span> F)</a>
<a class="sourceLine" id="cb19-4" data-line-number="4">cuts =<span class="st"> </span><span class="kw">cut</span>(hist<span class="op">$</span>breaks, <span class="kw">c</span>(<span class="op">-</span><span class="ot">Inf</span>,avg.count.cutoff , <span class="ot">Inf</span>))</a>
<a class="sourceLine" id="cb19-5" data-line-number="5"><span class="kw">plot</span>(hist, <span class="dt">col=</span><span class="kw">c</span>(<span class="st">&#39;red&#39;</span>, <span class="st">&#39;green&#39;</span>)[cuts],</a>
<a class="sourceLine" id="cb19-6" data-line-number="6">     <span class="dt">xlab=</span><span class="kw">expression</span>(Log[<span class="dv">2</span>]<span class="op">~</span><span class="st">&quot;average count of genes&quot;</span>),  <span class="dt">main=</span><span class="kw">paste0</span>(<span class="st">&quot; Histogram of mean expression. </span><span class="ch">\n</span><span class="st"> The green are kept: &quot;</span>,<span class="kw">sum</span>(log2.raw<span class="op">&gt;</span>avg.count.cutoff), <span class="st">&#39; out of &#39;</span>, <span class="kw">length</span>(log2.raw) ))</a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:geneqc1"></span>
<img src="scMTseq_of_AML_cells_files/figure-html/geneqc1-1.png" alt="Histogram of log mean counts for all genes - red shows mean log expression below the assigned cutoff" width="672" />
<p class="caption">
Figure 1.2: Histogram of log mean counts for all genes - red shows mean log expression below the assigned cutoff
</p>
</div>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1"><span class="co">## remove lowly expressed genes (&lt;2; log2 &lt;1)</span></a>
<a class="sourceLine" id="cb20-2" data-line-number="2">to.keep =<span class="st"> </span>log2.raw <span class="op">&gt;</span><span class="st"> </span>avg.count.cutoff</a>
<a class="sourceLine" id="cb20-3" data-line-number="3">sce.rna =<span class="st"> </span>sce.rna[to.keep,]</a></code></pre></div>
<p>11703 genes passed filtering.</p>
</div>
</div>
</div>
<div id="normalization" class="section level2">
<h2><span class="header-section-number">1.3</span> Normalization</h2>
<div id="scran" class="section level3">
<h3><span class="header-section-number">1.3.1</span> scran</h3>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1"><span class="co">## use scran to pool cells based on gene expression similarity and calculate size factors</span></a>
<a class="sourceLine" id="cb21-2" data-line-number="2"><span class="kw">system.time</span>({</a>
<a class="sourceLine" id="cb21-3" data-line-number="3">  <span class="co">##    user  system elapsed </span></a>
<a class="sourceLine" id="cb21-4" data-line-number="4">   <span class="co">## 0.366   0.027   0.435 </span></a>
<a class="sourceLine" id="cb21-5" data-line-number="5">sce.rna.scran =<span class="st"> </span><span class="kw">computeSumFactors</span>(sce.rna)</a>
<a class="sourceLine" id="cb21-6" data-line-number="6"><span class="co">## normalise using calculated size factors</span></a>
<a class="sourceLine" id="cb21-7" data-line-number="7">sce.rna.scran =<span class="st"> </span><span class="kw">normalize</span>(sce.rna.scran)</a>
<a class="sourceLine" id="cb21-8" data-line-number="8"><span class="kw">summary</span>(<span class="kw">sizeFactors</span>(sce.rna.scran))</a>
<a class="sourceLine" id="cb21-9" data-line-number="9"><span class="kw">plot</span>(<span class="kw">sizeFactors</span>(sce.rna.scran), sce.rna<span class="op">$</span>total_counts<span class="op">/</span><span class="fl">1e3</span>, <span class="dt">log=</span><span class="st">&quot;xy&quot;</span>, <span class="dt">pch=</span><span class="dv">16</span>, <span class="dt">col=</span><span class="st">&#39;purple&#39;</span>,</a>
<a class="sourceLine" id="cb21-10" data-line-number="10">    <span class="dt">ylab=</span><span class="st">&quot;Library size (thousands)&quot;</span>, <span class="dt">xlab=</span><span class="st">&quot;Size factor&quot;</span>, <span class="dt">main =</span> <span class="st">&quot;Scran Size Factors&quot;</span>)</a>
<a class="sourceLine" id="cb21-11" data-line-number="11">})</a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:normplotscran"></span>
<img src="scMTseq_of_AML_cells_files/figure-html/normplotscran-1.png" alt="Size factors from deconvolution, plotted against library sizes for all cells. Axes are shown on a log-scale. Since we have similar cell types (and not much DE genes) we expect to see less scatter around the trend line, whcih is the case in this plot." width="672" />
<p class="caption">
Figure 1.3: Size factors from deconvolution, plotted against library sizes for all cells. Axes are shown on a log-scale. Since we have similar cell types (and not much DE genes) we expect to see less scatter around the trend line, whcih is the case in this plot.
</p>
</div>
<pre><code>##    user  system elapsed 
##   0.297   0.011   0.322</code></pre>
</div>
<div id="tmm-edger" class="section level3">
<h3><span class="header-section-number">1.3.2</span> TMM (edgeR)</h3>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" data-line-number="1"><span class="co">## TMM</span></a>
<a class="sourceLine" id="cb23-2" data-line-number="2"><span class="co">## create singlecellexperiment object with edgeR normalisation factors</span></a>
<a class="sourceLine" id="cb23-3" data-line-number="3">sce.rna.TMM =<span class="st"> </span>sce.rna</a>
<a class="sourceLine" id="cb23-4" data-line-number="4"><span class="kw">sizeFactors</span>(sce.rna.TMM) =<span class="st"> </span><span class="kw">calcNormFactors</span>(<span class="kw">counts</span>(sce.rna.TMM))</a>
<a class="sourceLine" id="cb23-5" data-line-number="5">sce.rna.TMM =<span class="st"> </span><span class="kw">normalize</span>(sce.rna.TMM)</a>
<a class="sourceLine" id="cb23-6" data-line-number="6"><span class="kw">summary</span>(<span class="kw">sizeFactors</span>(sce.rna.TMM))</a></code></pre></div>
<pre><code>##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##  0.8867  0.9653  0.9968  1.0000  1.0365  1.1381</code></pre>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" data-line-number="1"><span class="kw">plot</span>(<span class="kw">sizeFactors</span>(sce.rna.TMM), sce.rna<span class="op">$</span>total_counts<span class="op">/</span><span class="fl">1e3</span>, <span class="dt">log=</span><span class="st">&quot;xy&quot;</span>, <span class="dt">pch=</span><span class="dv">16</span>, <span class="dt">col=</span><span class="st">&#39;purple&#39;</span>,</a>
<a class="sourceLine" id="cb25-2" data-line-number="2">    <span class="dt">ylab=</span><span class="st">&quot;Library size (thousands)&quot;</span>, <span class="dt">xlab=</span><span class="st">&quot;Size factor&quot;</span>, <span class="dt">main =</span> <span class="st">&quot;TMM Size Factors&quot;</span>)</a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:normplotTMM"></span>
<img src="scMTseq_of_AML_cells_files/figure-html/normplotTMM-1.png" alt="Size factors from TMM method. A trend line is not apparent." width="672" />
<p class="caption">
Figure 1.4: Size factors from TMM method. A trend line is not apparent.
</p>
</div>
</div>
<div id="deseq2" class="section level3">
<h3><span class="header-section-number">1.3.3</span> DESeq2</h3>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" data-line-number="1">sce.rna.DESeq2 =<span class="st"> </span>sce.rna</a>
<a class="sourceLine" id="cb26-2" data-line-number="2"><span class="kw">sizeFactors</span>(sce.rna.DESeq2) =<span class="st"> </span><span class="kw">estimateSizeFactorsForMatrix</span>(<span class="kw">counts</span>(sce.rna.DESeq2))</a>
<a class="sourceLine" id="cb26-3" data-line-number="3"><span class="co">## normalize using scater</span></a>
<a class="sourceLine" id="cb26-4" data-line-number="4">sce.rna.DESeq2 =<span class="st"> </span><span class="kw">normalize</span>(sce.rna.DESeq2)</a>
<a class="sourceLine" id="cb26-5" data-line-number="5"><span class="kw">summary</span>(<span class="kw">sizeFactors</span>(sce.rna.DESeq2))</a></code></pre></div>
<pre><code>##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##  0.4731  0.8474  0.9542  1.0000  1.1308  1.7616</code></pre>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" data-line-number="1"><span class="kw">plot</span>(<span class="kw">sizeFactors</span>(sce.rna.DESeq2), sce.rna<span class="op">$</span>total_counts<span class="op">/</span><span class="fl">1e3</span>, <span class="dt">log=</span><span class="st">&quot;xy&quot;</span>, <span class="dt">pch=</span><span class="dv">16</span>, <span class="dt">col =</span> <span class="st">&#39;purple&#39;</span>,</a>
<a class="sourceLine" id="cb28-2" data-line-number="2">    <span class="dt">ylab=</span><span class="st">&quot;Library size (thousands)&quot;</span>, <span class="dt">xlab=</span><span class="st">&quot;Size factor&quot;</span>, <span class="dt">main =</span> <span class="st">&quot;DESeq2 Size Factors&quot;</span>)</a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:DESeq2"></span>
<img src="scMTseq_of_AML_cells_files/figure-html/DESeq2-1.png" alt="Size factors from DESeq2 method." width="672" />
<p class="caption">
Figure 1.5: Size factors from DESeq2 method.
</p>
</div>
</div>
</div>
<div id="dimension-reduction" class="section level2">
<h2><span class="header-section-number">1.4</span> Dimension reduction</h2>
<p>We will use <em>Scran</em> and <em>DESeq2</em> normalisation for dimension reduction.</p>
<div id="pca" class="section level3">
<h3><span class="header-section-number">1.4.1</span> PCA</h3>
<div id="scran-1" class="section level4">
<h4><span class="header-section-number">1.4.1.1</span> scran</h4>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb29-1" data-line-number="1"><span class="co">## scran</span></a>
<a class="sourceLine" id="cb29-2" data-line-number="2">pca.scran =<span class="st"> </span><span class="kw">pca</span>(<span class="kw">t</span>(<span class="kw">logcounts</span>(sce.rna.scran)), <span class="dt">ncomp =</span> <span class="dv">5</span>)</a>
<a class="sourceLine" id="cb29-3" data-line-number="3"><span class="kw">source</span>(<span class="st">&#39;utils/plot.pca.R&#39;</span>)</a>
<a class="sourceLine" id="cb29-4" data-line-number="4"><span class="kw">plot.pca</span>(pca.scran, <span class="dt">top =</span> <span class="st">&#39;scran-normalised&#39;</span>)</a></code></pre></div>
<p><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-8-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<div id="deseq2-1" class="section level4">
<h4><span class="header-section-number">1.4.1.2</span> DESeq2</h4>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb30-1" data-line-number="1"><span class="co">## DESeq2</span></a>
<a class="sourceLine" id="cb30-2" data-line-number="2">pca.DESeq2 =<span class="st"> </span><span class="kw">pca</span>(<span class="kw">t</span>(<span class="kw">logcounts</span>(sce.rna.DESeq2)), <span class="dt">ncomp =</span> <span class="dv">5</span>)</a>
<a class="sourceLine" id="cb30-3" data-line-number="3"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">2</span>))</a>
<a class="sourceLine" id="cb30-4" data-line-number="4"><span class="kw">plot.pca</span>(pca.DESeq2, <span class="dt">top =</span> <span class="st">&#39;DESeq2-normalised&#39;</span>)</a></code></pre></div>
<p><img src="scMTseq_of_AML_cells_files/figure-html/unnamed-chunk-9-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="hierarchical-clustering" class="section level3">
<h3><span class="header-section-number">1.4.2</span> Hierarchical Clustering</h3>
<div id="hvgs" class="section level4">
<h4><span class="header-section-number">1.4.2.1</span> HVGs</h4>
<p>HVGs are chosen based on highest biological CVs which have a minimum mean expression specified</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb31-1" data-line-number="1"><span class="kw">source</span>(<span class="st">&#39;utils/calc.cv.R&#39;</span>)</a>
<a class="sourceLine" id="cb31-2" data-line-number="2">n.hvgs =<span class="st"> </span><span class="dv">500</span> <span class="co">## number of scRNAseq HVGs to keep for clustering</span></a>
<a class="sourceLine" id="cb31-3" data-line-number="3"><span class="co">## get a matrix of every gene&#39;s mean expression and BCV and order genes by BCV and plot BCV vs mean expression; order HVGs</span></a>
<a class="sourceLine" id="cb31-4" data-line-number="4">scran.gene.cv =<span class="st"> </span><span class="kw">calc.cv</span>(<span class="kw">logcounts</span>(sce.rna.scran), <span class="dt">mean.expr.cutoff =</span> <span class="dv">5</span>, <span class="dt">n.hvgs =</span> n.hvgs)</a>
<a class="sourceLine" id="cb31-5" data-line-number="5"><span class="co">## plot mean expression vs sd and cv</span></a>
<a class="sourceLine" id="cb31-6" data-line-number="6">scran.gene.cv<span class="op">$</span>plot.cv</a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:hvgs"></span>
<img src="scMTseq_of_AML_cells_files/figure-html/hvgs-1.png" alt="SD and biological CV against average counts (mean expression)" width="672" />
<p class="caption">
Figure 1.6: SD and biological CV against average counts (mean expression)
</p>
</div>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb32-1" data-line-number="1">hvgs =<span class="st"> </span><span class="kw">rownames</span>(scran.gene.cv<span class="op">$</span>expr.smry)[<span class="dv">1</span><span class="op">:</span>n.hvgs]</a>
<a class="sourceLine" id="cb32-2" data-line-number="2"><span class="co">## reduce sce object</span></a>
<a class="sourceLine" id="cb32-3" data-line-number="3">sce.rna.scran.hvg =<span class="st"> </span>sce.rna.scran[hvgs,]</a></code></pre></div>
</div>
<div id="euclidean-distance" class="section level4">
<h4><span class="header-section-number">1.4.2.2</span> Euclidean Distance</h4>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb33-1" data-line-number="1"><span class="kw">plotHeatmap</span>(sce.rna.scran.hvg, <span class="dt">features =</span> <span class="kw">rownames</span>(sce.rna.scran.hvg))</a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:heatmap"></span>
<img src="scMTseq_of_AML_cells_files/figure-html/heatmap-1.png" alt="Hierarchical clustering using highly variable genes - Euclidean Distance" width="70%" />
<p class="caption">
Figure 1.7: Hierarchical clustering using highly variable genes - Euclidean Distance
</p>
</div>
</div>
<div id="pearsons-r" class="section level4">
<h4><span class="header-section-number">1.4.2.3</span> Pearson’s <em>r</em></h4>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb34-1" data-line-number="1"><span class="kw">plotHeatmap</span>(sce.rna.scran.hvg, <span class="dt">features =</span> <span class="kw">rownames</span>(sce.rna.scran.hvg), <span class="dt">clutering_distance_rows =</span> <span class="st">&#39;correlation&#39;</span>, <span class="dt">clutering_distance_cols =</span> <span class="st">&#39;correlation&#39;</span>)</a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:heatmap-corr"></span>
<img src="scMTseq_of_AML_cells_files/figure-html/heatmap-corr-1.png" alt="Hierarchical clustering using highly variable genes - Pearson's r" width="70%" />
<p class="caption">
Figure 1.8: Hierarchical clustering using highly variable genes - Pearson’s r
</p>
</div>
</div>
</div>
<div id="tsne" class="section level3">
<h3><span class="header-section-number">1.4.3</span> tSNE</h3>
<p>For reduced scran-normaised matrix. We get some clusters when perplexity is really low:</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb35-1" data-line-number="1"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</a>
<a class="sourceLine" id="cb35-2" data-line-number="2"><span class="kw">set.seed</span>(<span class="dv">12321</span>)</a>
<a class="sourceLine" id="cb35-3" data-line-number="3">scran.tsne =<span class="st"> </span><span class="kw">Rtsne</span>(<span class="kw">t</span>(<span class="kw">logcounts</span>(sce.rna.scran.hvg)), <span class="dt">perplexity =</span> <span class="dv">2</span>)</a>
<a class="sourceLine" id="cb35-4" data-line-number="4"><span class="kw">plot</span>(scran.tsne<span class="op">$</span>Y, <span class="dt">main =</span> <span class="st">&#39;tSNE - Perplexity = 2&#39;</span>, <span class="dt">pch=</span><span class="dv">16</span>, <span class="dt">col=</span><span class="st">&#39;blue&#39;</span>, <span class="dt">xlab =</span> <span class="st">&#39;tSNE-1&#39;</span>, <span class="dt">ylab=</span><span class="st">&#39;tSNE-2&#39;</span>)</a>
<a class="sourceLine" id="cb35-5" data-line-number="5"><span class="kw">set.seed</span>(<span class="dv">12321</span>)</a>
<a class="sourceLine" id="cb35-6" data-line-number="6">scran.tsne =<span class="st"> </span><span class="kw">Rtsne</span>(<span class="kw">t</span>(<span class="kw">logcounts</span>(sce.rna.scran.hvg)), <span class="dt">perplexity =</span> <span class="dv">8</span>)</a>
<a class="sourceLine" id="cb35-7" data-line-number="7"><span class="kw">plot</span>(scran.tsne<span class="op">$</span>Y, <span class="dt">main =</span> <span class="st">&#39;tSNE - Perplexity = 8&#39;</span>, <span class="dt">pch=</span><span class="dv">16</span>, <span class="dt">col=</span><span class="st">&#39;blue&#39;</span>, <span class="dt">xlab =</span> <span class="st">&#39;tSNE-1&#39;</span>, <span class="dt">ylab=</span><span class="st">&#39;&#39;</span>)</a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:tSNE"></span>
<img src="scMTseq_of_AML_cells_files/figure-html/tSNE-1.png" alt="tSNE plots using HVGs" width="100%" />
<p class="caption">
Figure 1.9: tSNE plots using HVGs
</p>
</div>
</div>
<div id="looking-at-umap" class="section level3">
<h3><span class="header-section-number">1.4.4</span> Looking at UMAP</h3>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb36-1" data-line-number="1"><span class="kw">library</span>(umap)</a>
<a class="sourceLine" id="cb36-2" data-line-number="2"><span class="co">## UMAP on scran-normalised</span></a>
<a class="sourceLine" id="cb36-3" data-line-number="3">umap.scran.hvg =<span class="st"> </span><span class="kw">umap</span>(<span class="kw">t</span>(<span class="kw">logcounts</span>(sce.rna.scran.hvg)))</a>
<a class="sourceLine" id="cb36-4" data-line-number="4"><span class="kw">plot</span>(umap.scran.hvg<span class="op">$</span>layout, <span class="dt">main =</span> <span class="st">&#39;UMAP using HVGs&#39;</span>, <span class="dt">pch=</span><span class="dv">16</span>, <span class="dt">col=</span><span class="st">&#39;blue&#39;</span>, <span class="dt">xlab =</span> <span class="st">&#39;umap-1&#39;</span>, <span class="dt">ylab=</span><span class="st">&#39;umap-2&#39;</span>)</a></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:UMAP"></span>
<img src="scMTseq_of_AML_cells_files/figure-html/UMAP-1.png" alt="UMAP plot using HVGs" width="672" />
<p class="caption">
Figure 1.10: UMAP plot using HVGs
</p>
</div>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="scbs-seq-data.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
